---
title: "Metagenomic_reads"
author: "Sharok"
date: "4/28/2023"
output: html_document
---

```{r}

library(tidyverse)
library(tidytext)

```

```{r}

mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

```

## Marker-level analysis
Reading marker abundance files, parse them into one
```{r}

path="../data/Metaphlan4/marker_abund"
#outpath="metaphlan4"
data.frame(sample.id = paste(dir(path,
                                 pattern = "*_abund_norm.txt"))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read.csv(file.path(path, .),
                                                 sep = "\t", header = F,
                                                 comment.char = "#"))) %>%
unnest() %>%
mutate(sample.id= gsub(".marker_abund_norm.txt", "", sample.id)) %>%
  spread(sample.id, V2) %>%
  rename(Marker="V1") %>%
  mutate(temp=Marker) %>%
  separate(temp, c("sp_marker", "temp1", "temp2"), sep = "__") %>%
  mutate(st_marker= case_when(!is.na(temp2) ~ paste(temp1, temp2, sep = "__"),
                              TRUE ~ paste(temp1))) %>%
  select(sp_marker, st_marker, Marker, DonorB_D_2013:PMCL883) -> marker_lvl

# sum of abundance based on normalized data (will be used for relative abund)
marker_lvl %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker, na.rm = TRUE) %>%
  ungroup() %>%
  group_by(sample) %>%
  summarise(total_abund=sum(abundance)) -> total.abund

# convert the marker-level data to relative abundance
# this is just to use a standard cut-off
marker_lvl %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker, na.rm = TRUE) %>%
  left_join(total.abund) %>%
  mutate(relative_abund= abundance/total_abund * 100) %>%
  select(Marker, sample, relative_abund) %>%
  spread(sample, relative_abund) %>%
  mutate(temp=Marker) %>%
  separate(temp, c("sp_marker", "temp1", "temp2"), sep = "__") %>%
  mutate(st_marker= case_when(!is.na(temp2) ~ paste(temp1, temp2, sep = "__"),
                              TRUE ~ paste(temp1))) %>%
  select(-temp1, -temp2) -> marker_lvl_RA

```


```{r}

################################################################################
# a function to detect common engraftment based on strains (metaphlan markers)
################################################################################
stCE <- function(marker_lvl, mapfile, strains, cutoff=cutoff, ...){
  # convert marker-lvl to long format, make Study_ID variable (in mapfile),
  # and select only donor's strains
  marker_lvl %>%
  filter(Marker %in% strains) %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_markerlvl
  # complete list of engrafted
  long_markerlvl %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(Marker, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -Marker) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i),
                             n = length(Emark.1[[i]]$Marker))
  }
  do.call(rbind.data.frame, Emark.2)
}

################################################################################

stCE_2 <- function(marker_lvl, mapfile, strains, cutoff=cutoff, ...){
  # convert marker-lvl to long format, make Study_ID variable (in mapfile),
  # and select only donor's strains
  marker_lvl %>%
  filter(Marker %in% strains) %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_markerlvl
  # complete list of engrafted
  long_markerlvl %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(Marker, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -Marker) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i), Emark.1[[i]])
  }
  common_engraft <- do.call(rbind.data.frame, Emark.2)
  common_engraft %>%
    gather(Study_ID, abundance, -Marker, -Npts) %>%
    filter(abundance >= cutoff) %>%
    left_join(mapfile %>% select(Study_ID, Fig_lab, Treatment, Remission) %>%
              distinct()) %>%
    group_by(Npts, Marker, Treatment, Remission) %>%
    tally() %>%
    mutate(cat= paste(Treatment, Remission, sep = "")) %>%
    ungroup() %>%
    select(Npts, Marker, n, cat) %>%
    spread(cat, n) %>%
    mutate(M_cat= case_when(is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTNoRes) &
                               FMTRes > 0 ~ paste("FMT-Res"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTRes) &
                               FMTNoRes > 0 ~ paste("FMT-NoRes"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) ~ paste("FMT"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboRes) &
                               PlaceboNoRes > 0 ~ paste("Placebo-NoRes"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboNoRes) &
                               PlaceboRes > 0 ~ paste("Placebo-Res"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) ~ paste("Placebo"),
                             TRUE ~ paste("Both"))) %>%
  mutate(across(everything(), ~replace_na(., 0)))
}

################################################################################

```



## engraftment cut-off for strains

```{r}

marker_lvl_RA %>%
  gather(sample, abundance, -sp_marker, -st_marker, -Marker) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) %>%
  mutate(sample= gsub("DonorB_D_", "", sample)) -> donorB_markers


# a function to filter marker above cutoff:
Fpos_finder <- function(table, cutoff){
  #table %>%
  #  ungroup() %>%
  #  group_by(sample) %>%
  #  summarise(total_abund=sum(abundance))-> total.abund
  
  # total number of unqiue marker in each sample
  table %>%
    group_by(sample) %>%
    summarise(total=n()) -> total.marker
  
  # total number of unique marker above cutoff
  table %>%
    #left_join(total.abund) %>%
    #mutate(relative_abund= abundance/total_abund * 100) %>%
    mutate(relative_abund= abundance) %>%
    filter(relative_abund >= cutoff) %>%
    group_by(sample) %>%
    summarise(pos_n=n()) %>%
    left_join(total.marker) %>%
    mutate(pos_p=pos_n/total * 100) %>%
    mutate(cutoff=paste(cutoff))
}

cols <- c("2013" = "#bae4b3",
          "2016" = "#74c476", 
          "May17" = "#31a354",
          "Oct17" = "#006d2c")

# comparison of detection cut-offs
Fpos_visualizer <- function(table) {
  
  vector <- c(0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1, 1)
  vector <- as.character(vector)  # Convert vector to character for proper factor levels
  
  res <- list()
  for (i in vector) {
    res[[i]] <- Fpos_finder(table, as.numeric(i))
  }
  
  tbl <- do.call(rbind.data.frame, res)
  tbl$cutoff <- factor(tbl$cutoff, levels = vector)
  
  ggplot(tbl, aes(cutoff, pos_p, color = sample, fill=sample)) +
    geom_line(aes(group = sample)) +
    scale_color_manual(values = cols) +
    scale_fill_manual(values = cols) +
    theme_bw() +
    theme(legend.position = c(0.8,0.7),
          legend.title = element_blank()) +
    xlab("Relative abundance (%)") +
    ylab("% of unique strain marker")
}

p1 <- Fpos_visualizer(donorB_markers)

p2 <- donorB_markers %>%
  ungroup() %>%
  group_by(sample) %>%
  summarise(percent=abundance/sum(abundance) * 100) %>%
ggplot(aes(x=percent, fill=sample, color=sample)) +
    geom_histogram(position="identity", colour="grey40", alpha=0.7, bins = 10) +
    scale_x_log10() +
    facet_grid(. ~ sample) +
    geom_vline(xintercept = 0.0001, color = "red") +
    scale_fill_manual(values = cols) +
    theme_classic() +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 90)) +
  xlab("Relative abundance (%)") +
  ylab("# of strain markers")

cowplot::plot_grid(p1, p2, ncol = 1)
ggsave("../figs/reads_strain_cutoff.png",
       width = 14, height = 14, units = "cm")

```

## donorB markers and set engraftment cutoff

```{r}

marker_lvl_RA %>%
  gather(sample, abundance, -sp_marker, -st_marker, -Marker) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) %>%
  mutate(sample= gsub("DonorB_D_", "", sample)) -> donorB_markers

##### seting engraftment cut-off
# based on donor B benchmarking data: 0.0001 (as shown above)
cutoff=0.0001
####################################################################

# selecting only those donor B markers that meet cutoff:
# > average of minimum detections
donorB_markers %>%
  filter(as.numeric(abundance) > cutoff) -> donorB_markers

#####################################################################
# selecting only donor B markers that present across all donor B samples:
#donorB_markers %>%
#  spread(sample, abundance, fill = 0) %>%
#  select("Marker", "2013":Oct17) -> B_wide
#B_wide <- B_wide[rowSums(B_wide[,2:ncol(B_wide)] > cutoff) >= 2, ]
#B_wide %>%
#  gather(sample, abundance, -Marker) -> donorB_markers
  
#####################################################################

donorB_markers %>%
  select(Marker) %>% distinct() %>%
  pull() -> B_markers

```


## Permutation on treatments for strins
1k permutation with no replace was done on treatment data
then the area under the curve were calculated for FMT and Placebo
see ../script/permutation_strains.R for more details

```{r}

# load data:

auc_null<- read.csv("../permut_data/strains_auc_null.txt",
                    sep = "\t")
final_result<- read.csv("../permut_data/strains_permutated.txt",
                    sep = "\t")


```

## merging permutation result with observed

```{r}

# Get the observed FMT values
stCE(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment %in% c("FMT")) %>%
  mutate(Treatment=paste("withinFMT(n=12)"))-> C2
#3. WithinPlacebo
stCE(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment %in% c("Placebo")) %>%
  mutate(Treatment=paste("withinPlacebo(n=12)"))-> C3

# Get the observed FMT and Placebo AUC values and find the difference
fmt_auc = integrate(approxfun(C2$Npts, C2$n), 1, as.numeric(max(C2$Npts)))
plc_auc = integrate(approxfun(C3$Npts, C3$n), 1, as.numeric(max(C3$Npts)))
obsdeltaAuc = fmt_auc$value - plc_auc$value

ggplot(data.frame(auc_null), aes(x = deltaAUC)) +
    geom_density() +
    geom_vline(xintercept = obsdeltaAuc)

# Place that difference on the null distribution of differences and get its p-value
quant_fun = ecdf(auc_null[,1])
auc_pval = 1 - quant_fun(obsdeltaAuc)
auc_pval

# merging the results together:
plt_df = rbind(C2, C3)
ribdat = plt_df %>% spread(Treatment, n)

cols <- c("withinFMT(n=12)" = "#1b7837",
          "withinPlacebo(n=12)" = "#999999")

f1 <- ggplot(plt_df, aes(as.numeric(Npts), n, color=Treatment)) +
  geom_point() + geom_line() +
  geom_ribbon(data = ribdat, aes(ymin=`withinPlacebo(n=12)`,
                                 ymax=`withinFMT(n=12)`, 
                                 x = as.numeric(Npts)),
              linetype=2, alpha=0.1, inherit.aes = FALSE)+
  annotate("text", x=9, y=12000, size=2,
           label= paste("Permutation test,\n p=", auc_pval, sep = " ")) +
  #scale_y_log10() +
  scale_x_continuous(breaks = seq(0,max(as.numeric(plt_df$Npts)) + 1, 2)) +
  scale_color_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.title = element_blank())

```


## common engraftment - strains

```{r}


#1. total:
stCE(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment %in% c("FMT", "Placebo")) %>%
  mutate(Type=paste("Total(n=24)"))-> C1


#1. Unique to FMT or Placebo and FMT responder vs non-
stCE_2(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment %in% c("FMT", "Placebo")) -> U
U %>%
  group_by(Npts, M_cat) %>%
  tally() %>%
  mutate(Type= case_when(str_detect(M_cat, "FMT") ~ paste("uniquetoFMT"),
                           str_detect(M_cat, "Placebo") ~ paste("uniquetoPlacebo"),
                           TRUE ~ paste("Overlap"))) %>%
  ungroup() %>% group_by(Npts, Type) %>%
  summarise(n=sum(n)) -> U1

U %>%
  group_by(Npts, M_cat) %>%
  tally() %>%
  filter(str_detect(M_cat, "FMT")) %>%
  mutate(Type= case_when(str_detect(M_cat, "FMT-Res") ~ paste("uniquetoFMT-Res"),
                           str_detect(M_cat, "FMT-NoRes") ~ paste("uniquetoFMT-NoRes"),
                           TRUE ~ paste("uniquetoFMT-Overlap"))) %>%
  select(Npts, Type, n)-> U2

# merging the results together:
rbind(C1, C2, C3) -> tbl1
rbind(U1, U2) -> tbl2

cols <- c("Total(n=24)" = "black",
          "withinFMT(n=12)" = "#1b7837",
          "withinPlacebo(n=12)" = "#999999",
          "withinFMT-Res(n=6)" = "#2166ac",
          "withinFMT-NoRes(n=6)" = "#b2182b",
          "Overlap" = "#fc8d59",
          "uniquetoFMT" = "#1b7837",
          "uniquetoPlacebo" = "#999999",
          "uniquetoFMT-Res" = "#2166ac",
          "uniquetoFMT-NoRes" = "#b2182b",
          "uniquetoFMT-Overlap" = "#6A3F6C")


f2 <- ggplot(U1, aes(as.numeric(Npts), as.numeric(n), fill=Type)) +
  #geom_point() + geom_line(aes(group=Type)) +
  scale_color_manual(values = cols) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = cols) +
  #scale_y_log10() +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_blank()) 

U2$Type <- factor(U2$Type, levels = c("uniquetoFMT-Overlap",
                                      "uniquetoFMT-NoRes",
                                      "uniquetoFMT-Res"))
f3 <- ggplot(U2, aes(as.numeric(Npts), as.numeric(n), fill=Type)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_blank()) #+
   #xlab("Increasting # of patients") + ylab("Engrafted strains")
cowplot::plot_grid(f1, f2, f3, ncol = 1, hjust = T, vjust = T)
ggsave("../figs/reads_CESt.png",
       width = 6, height = 12, units = "cm")
write.table(tbl1, "../figs/tbl1.txt", 
            sep = "\t", row.names = F, quote = F)
write.table(U1, "../figs/U1.txt", 
            sep = "\t", row.names = F, quote = F)
write.table(U2, "../figs/U2.txt",
            sep = "\t", row.names = F, quote = F)
```



```{r}

# Sample data frame
data <- data.frame(
  Fig_lab = c("pt1", "pt2", "pt3", "pt4", "pt5", "pt6"),
  Treatment = c("FMT", "Placebo", "FMT", "Placebo", "FMT", "Placebo")
)

# Function that operates on a data frame
custom_function <- function(df) {
  df %>%
    group_by(Treatment) %>%
    tally()
}

# Number of permutations
num_permutations <- 10

# Create an empty list to store the results of each iteration
iteration_results <- list()

# Loop through permutations
for (i in 1:num_permutations) {
  # Permute the Treatment column
  permuted_data <- data %>%
    mutate(Treatment = sample(Treatment, replace = TRUE))
  
  # Store the result in the iteration_results list
  iteration_results[[i]] <- custom_function(permuted_data)
  iteration_results[[i]]$it <- paste("it", i, sep="_")
}

# Combine the iteration results into a single data frame
final_result <- do.call(rbind, iteration_results)

# Print the final result
print(final_result)

```




## commonly engrafted strains for 6pts

```{r}

################################################################################
# a function to detect common engraftment in 3pts 
################################################################################
stCE_3pts <- function(marker_lvl, mapfile, strains, cutoff=cutoff, ...){
  # convert marker-lvl to long format, make Study_ID variable (in mapfile),
  # and select only donor's strains
  marker_lvl %>%
  filter(Marker %in% strains) %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_markerlvl
  # complete list of engrafted
  long_markerlvl %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(Marker, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -Marker) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  
  engraft_3 <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= 6, ]
  engraft_3 %>%
    gather(Study_ID, abundance, -Marker) %>%
    filter(abundance >= cutoff)
  
}

################################################################################
################################################################################

##### seting engraftment cut-off
# based on donor B benchmarking data: 0.0001 (as shown above)
cutoff=0.0001

######### common engraftment for 3 in FMT ###############################
stCE_3pts(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "FMT") -> commonE_FMT_3
######### common engraftment for 3 in Placebo ############################
stCE_3pts(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "Placebo") -> commonE_Placebo_3

stCE_3pts(marker_lvl_RA, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment %in% c("FMT", "Placebo")) -> tbl

tbl %>%
#commonE_FMT_3 %>%
#  bind_rows(commonE_Placebo_3) %>%
  left_join(mapfile %>% select(Study_ID, Fig_lab, Treatment, Remission) %>%
              distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> engraft_3pts

engraft_3pts %>%
  group_by(Marker, Treatment, Remission) %>%
  tally() %>%
  mutate(cat= paste(Treatment, Remission, sep = "")) %>%
  ungroup() %>%
  select(Marker, n, cat) %>%
  spread(cat, n) %>%
   mutate(M_cat= case_when(is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTNoRes) &
                               FMTRes > 0 ~ paste("FR"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTRes) &
                               FMTNoRes > 0 ~ paste("FN"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) ~ paste("F"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboRes) &
                               PlaceboNoRes > 0 ~ paste("PN"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboNoRes) &
                               PlaceboRes > 0 ~ paste("PR"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) ~ paste("P"),
                             TRUE ~ paste("Other"))) %>%
  mutate(across(everything(), ~replace_na(., 0))) %>%

  mutate(M_cat_order= case_when(M_cat == "F" ~ paste(as.numeric(FMTNoRes+FMTRes)),
                                   M_cat == "FR" ~ paste(as.numeric(FMTRes)),
                                   M_cat == "FN" ~ paste(as.numeric(FMTNoRes)),
                                   M_cat == "P" ~ paste(as.numeric(PlaceboRes,PlaceboNoRes)),
                                   M_cat == "PR" ~ paste(as.numeric(PlaceboRes)),
                                   M_cat == "PN" ~ paste(as.numeric(PlaceboNoRes)),
                                   M_cat == "Other" ~ paste(as.numeric(FMTRes+FMTNoRes+PlaceboNoRes+PlaceboRes)),
                                   TRUE ~ paste(0))) -> engraft_3pts_cat

engraft_3pts %>%
  left_join(engraft_3pts_cat) -> engraft_3pts_strain

ggplot(engraft_3pts_strain, aes(reorder_within(Marker, -as.numeric(M_cat_order), M_cat),
                               Fig_lab, fill=Treatment)) +
  geom_tile() +
  facet_grid(Treatment~M_cat, space = "free", scales = "free") +
  scale_fill_manual(values = c("FMT" = "#008837",
                               "Placebo" = "#bdbdbd")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_text(size=7),
        strip.text.x = element_blank(),
        legend.position = "none",
        axis.title.y = element_blank()) +
  xlab("Donor strains commonly engrafted in >=6 pts")
ggsave("../figs/reads_CESt_6pts.png", width = 20, height = 8, units = "cm")

unique(engraft_3pts_strain$M_cat)

engraft_3pts_strain %>%
  filter(M_cat == "PN") %>%
  select(Marker) %>% distinct() %>% pull() -> strains_of_interest

#the only bacteria unqiue to Placebo non-responder:
# "SGB6939__FFILHHMK_01448" which is a Veillonella parvula (SGB6939)

ggplot(engraft_3pts_strain %>%
         filter(M_cat == "F"), 
       aes(reorder_within(Marker, -as.numeric(M_cat_order), M_cat),
                               Fig_lab, fill=Treatment)) +
  geom_tile() +
  facet_grid(Treatment~M_cat, space = "free", scales = "free") +
  scale_fill_manual(values = c("FMT" = "#008837",
                               "Placebo" = "#bdbdbd")) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        axis.title.x = element_text(size=7),
        #strip.text.x = element_blank(),
        legend.position = "none",
        axis.title.y = element_blank()) +
  xlab("Donor strains commonly engrafted in >=6 pts")

engraft_3pts_strain %>%
  group_by(M_cat, Marker) %>%
  summarise(n=n()) %>%
  group_by(M_cat) %>%
  summarise(Markers=n()) %>%
  mutate(Treatment= case_when(str_detect(M_cat, "F") ~ paste("FMT"),
                              str_detect(M_cat, "P") ~ paste("Placebo"),
                              str_detect(M_cat, "Other") ~ paste("Both"))) %>%
  mutate(Treatment = factor(Treatment, levels = c("FMT", "Placebo", "Both"))) %>%
  ggplot(aes(Treatment, Markers)) +
  geom_bar(stat = "identity") +
  scale_y_sqrt() +
  theme_bw() +
  theme(axis.title.x = element_blank()) +
  ylab("# of unique markers engrafted in >= 6 patients")
ggsave("../figs/reads_CESt_6pts_bar.png",
       width = 8, height = 8, units = "cm")
  
```

## Species-level analysis
Reading metaphlan4 default output (merged)
relative abundance
```{r}

org <- read.csv("../data/Metaphlan4/merged_metaphlan4.txt", 
                sep = "\t", header = F, 
                comment.char = "#")
org %>%
  filter(V1 != "clade_name") %>%
  # selecting only at strain level- contains all lineage- avoid duplicates
  filter(grepl("t__S", V1)) -> metaphlan
colnames(metaphlan) <- gsub(".mp.profile_sp", "", org[1,])

metaphlan %>%
  mutate(temp=clade_name) %>%
  separate(temp, c("Kingdom", "Phylum", "Class", 
                       "Order", "Family", "Genus", 
                       "Species", "marker"), sep = "[|]") %>%
  select(clade_name,Kingdom:marker) %>%
  mutate(sp_marker= gsub("t__", "", marker)) %>%
  mutate(sp_marker=gsub("_group", "", sp_marker))-> lineage

```


```{r}

################################################################################
# a function to detect common engraftment based on species (metaphlan)
################################################################################
spCE <- function(metaphlan, mapfile, species, cutoff=cutoff, ...){
  # convert metaphlan to long format, make Study_ID variable (in mapfile),
  # and select only donor's species
  metaphlan %>%
  filter(clade_name %in% species) %>%
  gather(sample, abundance, -clade_name) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_metaphlan
  # complete list of engrafted
  long_metaphlan %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(clade_name, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -clade_name) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i),
                             n = length(Emark.1[[i]]$clade_name))
  }
  do.call(rbind.data.frame, Emark.2)
}
################################################################################

spCE_2 <- function(metaphlan, mapfile, species, cutoff=cutoff, ...){
  
metaphlan %>%
  filter(clade_name %in% species) %>%
  gather(sample, abundance, -clade_name) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_metaphlan
  # complete list of engrafted
  long_metaphlan %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(clade_name, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -clade_name) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i), Emark.1[[i]])
  }
  common_engraft <- do.call(rbind.data.frame, Emark.2)
  common_engraft %>%
    gather(Study_ID, abundance, -clade_name, -Npts) %>%
    filter(abundance >= cutoff) %>%
    left_join(mapfile %>% select(Study_ID, Fig_lab, Treatment, Remission) %>%
              distinct()) %>%
    group_by(Npts, clade_name, Treatment, Remission) %>%
    tally() %>%
    mutate(cat= paste(Treatment, Remission, sep = "")) %>%
    ungroup() %>%
    select(Npts, clade_name, n, cat) %>%
    spread(cat, n) %>%
    mutate(M_cat= case_when(is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTNoRes) &
                               FMTRes > 0 ~ paste("FMT-Res"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTRes) &
                               FMTNoRes > 0 ~ paste("FMT-NoRes"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) ~ paste("FMT"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboRes) &
                               PlaceboNoRes > 0 ~ paste("Placebo-NoRes"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboNoRes) &
                               PlaceboRes > 0 ~ paste("Placebo-Res"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) ~ paste("Placebo"),
                             TRUE ~ paste("Both"))) %>%
  mutate(across(everything(), ~replace_na(., 0)))
}


################################################################################

```

## engraftment cut-off for species

```{r}

metaphlan %>%
  gather(sample, abundance, -clade_name) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) %>%
  mutate(sample= gsub("DonorB_D_", "", sample)) -> donorB_species

# a function to filter species above cutoff:
Fpos_finder2 <- function(table, cutoff){
  # total number of unqiue species in each sample
  table %>%
    group_by(sample) %>%
    summarise(total=n()) -> total.species
  
  # total number of unique marker above cutoff
  table %>%
    mutate(relative_abund= abundance) %>%
    filter(relative_abund >= cutoff) %>%
    group_by(sample) %>%
    summarise(pos_n=n()) %>%
    left_join(total.species) %>%
    mutate(pos_p=pos_n/total * 100) %>%
    mutate(cutoff=paste(cutoff))
}

cols <- c("2013" = "#bae4b3",
          "2016" = "#74c476", 
          "May17" = "#31a354",
          "Oct17" = "#006d2c")

# comparison of detection cut-offs
Fpos_visualizer2 <- function(table) {
  
  vector <- c(0.001, 0.005, 0.01, 0.1, 1, 2, 3)
  vector <- as.character(vector)  # Convert vector to character for proper factor levels
  
  res <- list()
  for (i in vector) {
    res[[i]] <- Fpos_finder2(table, as.numeric(i))
  }
  
  tbl <- do.call(rbind.data.frame, res)
  tbl$cutoff <- factor(tbl$cutoff, levels = vector)
  
  ggplot(tbl, aes(cutoff, pos_p, color = sample, fill=sample)) +
    geom_line(aes(group = sample)) +
    scale_color_manual(values = cols) +
    scale_fill_manual(values = cols) +
    theme_bw() +
    theme(legend.position = c(0.8,0.7),
          legend.title = element_blank()) +
    xlab("Relative abundance (%)") +
    ylab("% of unique species")
}

p1 <- Fpos_visualizer2(donorB_species)

p2 <- donorB_species %>%
  ungroup() %>%
  group_by(sample) %>%
  summarise(percent=as.numeric(abundance)/sum(as.numeric(abundance)) * 100) %>%
ggplot(aes(x=percent, fill=sample, color=sample)) +
    geom_histogram(position="identity", colour="grey40", alpha=0.7, bins = 10) +
    scale_x_log10() +
    facet_grid(. ~ sample) +
    geom_vline(xintercept = 0.005, color = "red") +
    scale_fill_manual(values = cols) +
    theme_classic() +
    theme(legend.position = "none",
          axis.text.x = element_text(angle = 90)) +
  xlab("Relative abundance (%)") +
  ylab("# of species")

cowplot::plot_grid(p1, p2, ncol = 1)
ggsave("../figs/reads_species_cutoff.png",
       width = 14, height = 14, units = "cm")

```


```{r}


metaphlan %>%
  gather(sample, abundance, -clade_name) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) %>%
  mutate(sample= gsub("DonorB_D_", "", sample)) -> donorB_species

mapfile %>% select(Study_ID, Treatment, Timepoint) -> map

##### seting engraftment cut-off
# based on donor B benchmarking data: 0.005 (as shown above)
# minimum relative abundance detected maximum (top quartile) of species
cutoff= 0.005
######

# working with only donor B species that meeet cutoff
# > average of minimum detections
donorB_species %>%
  filter(as.numeric(abundance) > cutoff) -> donorB_species

#####################################################################
# selecting only donor B specoes that present across all donor B samples:
#donorB_species %>%
#  spread(sample, abundance, fill = 0) %>%
#  select("clade_name", "2013":Oct17) -> B_wide
#B_wide <- B_wide[rowSums(B_wide[,2:ncol(B_wide)] > cutoff) >= 2, ]
#B_wide %>%
#  gather(sample, abundance, -clade_name) -> donorB_species
  
#####################################################################

donorB_species %>%
  select(clade_name) %>% distinct() %>%
  pull() -> B_species

#1. total:
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment %in% c("FMT", "Placebo")) %>%
  mutate(Type=paste("Total(n=24)"))-> C1
#2 WithinFMT:
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment %in% c("FMT")) %>%
  mutate(Type=paste("withinFMT(n=12)"))-> C2
#3 WithinPlacebo:
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment %in% c("Placebo")) %>%
  mutate(Type=paste("withinPlacebo(n=12)"))-> C3
#4 WithtinFMTresponder:
spCE(metaphlan, mapfile, species = B_species, cutoff = cutoff,
     Treatment == "FMT" & Remission == "Res") %>%
     mutate(Type=paste("withinFMT-Res(n=6)"))-> C4
#5 WithtinFMTresponder:
spCE(metaphlan, mapfile, species = B_species, cutoff = cutoff,
     Treatment == "FMT" & Remission == "NoRes") %>%
     mutate(Type=paste("withinFMT-NoRes(n=6)"))-> C5

#1. Unique to FMT or Placebo and FMT responder vs non-
spCE_2(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment %in% c("FMT", "Placebo")) -> U

U %>%
  group_by(Npts, M_cat) %>%
  tally() %>%
  mutate(Type= case_when(str_detect(M_cat, "FMT") ~ paste("uniquetoFMT"),
                           str_detect(M_cat, "Placebo") ~ paste("uniquetoPlacebo"),
                           TRUE ~ paste("Overlap"))) %>%
  ungroup() %>% group_by(Npts, Type) %>%
  summarise(n=sum(n)) -> U1

U %>%
  group_by(Npts, M_cat) %>%
  tally() %>%
  filter(str_detect(M_cat, "FMT")) %>%
  mutate(Type= case_when(str_detect(M_cat, "FMT-Res") ~ paste("uniquetoFMT-Res"),
                           str_detect(M_cat, "FMT-NoRes") ~ paste("uniquetoFMT-NoRes"),
                           TRUE ~ paste("uniquetoFMT-Overlap"))) %>%
  select(Npts, Type, n)-> U2

# merging the results together:
rbind(C1, C2, C3) -> tbl1
rbind(U1, U2) -> tbl2

cols <- c("Total(n=24)" = "black",
          "withinFMT(n=12)" = "#1b7837",
          "withinPlacebo(n=12)" = "#999999",
          "withinFMT-Res(n=6)" = "#2166ac",
          "withinFMT-NoRes(n=6)" = "#b2182b",
          "Overlap" = "#fc8d59",
          "uniquetoFMT" = "#1b7837",
          "uniquetoPlacebo" = "#999999",
          "uniquetoFMT-Res" = "#2166ac",
          "uniquetoFMT-NoRes" = "#b2182b",
          "uniquetoFMT-Overlap" = "#6A3F6C")


f1 <- ggplot(tbl1, aes(as.numeric(Npts), n, color=Type)) +
  geom_point() + geom_line(aes(group=Type)) +
  scale_color_manual(values = cols) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "none",
        axis.title = element_blank()) #+
  #xlab("Increasting # of patients") + ylab("Engrafted species")

f2 <- ggplot(U1, aes(as.numeric(Npts), n, fill=Type)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = cols) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "none",
        axis.title = element_blank()) #+
   #xlab("Increasting # of patients") + ylab("Engrafted species")

f3 <- ggplot(U2, aes(as.numeric(Npts), n, color=Type)) +
  geom_point() + geom_line(aes(group=Type)) +
  scale_color_manual(values = cols) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "none",
        axis.title = element_blank()) #+
   #xlab("Increasting # of patients") + ylab("Engrafted species")
cowplot::plot_grid(f1, f2, f3, ncol = 1, hjust = T, vjust = T)
ggsave("../figs/reads_CESp.png",
       width = 6, height = 12, units = "cm")

```

## commonly engrafted species for 6pts

```{r}

################################################################################
# a function to detect common engraft species (metaphlan) in >= 3pts
################################################################################
spCE_3pts <- function(metaphlan, mapfile, species, cutoff=cutoff, ...){
  # convert metaphlan to long format, make Study_ID variable (in mapfile),
  # and select only donor's species
  metaphlan %>%
  filter(clade_name %in% species) %>%
  gather(sample, abundance, -clade_name) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_metaphlan
  # complete list of engrafted
  long_metaphlan %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(clade_name, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -clade_name) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  
  engraft_3 <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= 6, ]
  engraft_3 %>%
    gather(Study_ID, abundance, -clade_name) %>%
    filter(abundance >= cutoff)
}
################################################################################
################################################################################

##### seting engraftment cut-off
# based on donor B benchmarking data: 0.005 (as shown above)
# minimum relative abundance detected maximum (top quartile) of species
cutoff= 0.005
######

spCE_3pts(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment %in% c("FMT", "Placebo")) -> tbl

tbl %>%
  left_join(mapfile %>% select(Study_ID, Fig_lab, Treatment, Remission) %>%
              distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> engraft_3pts

engraft_3pts %>%
  group_by(clade_name, Treatment, Remission) %>%
  tally() %>%
  mutate(cat= paste(Treatment, Remission, sep = "")) %>%
  ungroup() %>%
  select(clade_name, n, cat) %>%
  spread(cat, n) %>%
   mutate(M_cat= case_when(is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTNoRes) &
                               FMTRes > 0 ~ paste("FR"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) &
                               is.na(FMTRes) &
                               FMTNoRes > 0 ~ paste("FN"),
                             
                             is.na(PlaceboRes) &
                               is.na(PlaceboNoRes) ~ paste("F"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboRes) &
                               PlaceboNoRes > 0 ~ paste("PN"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) &
                               is.na(PlaceboNoRes) &
                               PlaceboRes > 0 ~ paste("PR"),
                             
                             is.na(FMTRes) &
                               is.na(FMTNoRes) ~ paste("P"),
                             TRUE ~ paste("Other"))) %>%
  mutate(across(everything(), ~replace_na(., 0))) %>%

  mutate(M_cat_order= case_when(M_cat == "F" ~ paste(as.numeric(FMTNoRes+FMTRes)),
                                   M_cat == "FR" ~ paste(as.numeric(FMTRes)),
                                   M_cat == "FN" ~ paste(as.numeric(FMTNoRes)),
                                   M_cat == "P" ~ paste(as.numeric(PlaceboRes,PlaceboNoRes)),
                                   M_cat == "PR" ~ paste(as.numeric(PlaceboRes)),
                                   M_cat == "PN" ~ paste(as.numeric(PlaceboNoRes)),
                                   M_cat == "Other" ~ paste(as.numeric(FMTRes+FMTNoRes+PlaceboNoRes+PlaceboRes)),
                                   TRUE ~ paste(0))) -> engraft_3pts_cat

engraft_3pts %>%
  left_join(engraft_3pts_cat) -> engraft_3pts_species

ggplot(engraft_3pts_species, aes(reorder_within(clade_name, -as.numeric(M_cat_order), M_cat),
                               Fig_lab, fill=Treatment)) +
  geom_tile() +
  facet_grid(Treatment~M_cat, space = "free", scales = "free") +
  scale_fill_manual(values = c("FMT" = "#008837",
                               "Placebo" = "#bdbdbd")) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_text(size=7),
        strip.text.x = element_blank(),
        legend.position = "none",
        axis.title.y = element_blank()) +
  xlab("Donor species commonly engrafted in >=6 pts")
ggsave("../figs/reads_CESp_6pts.png", width = 20, height = 8, units = "cm")


unique(engraft_3pts_species$M_cat)

engraft_3pts_species %>%
  filter(M_cat == "P") %>%
  select(clade_name) %>% distinct() %>% pull() -> species_of_interest

#the only bacteria unqiue to Placebo:
# s__Veillonella_parvula|t__SGB6939"

engraft_3pts_species %>%
  group_by(M_cat, clade_name) %>%
  summarise(n=n()) %>%
  group_by(M_cat) %>%
  summarise(Species=n()) %>%
  mutate(Treatment= case_when(str_detect(M_cat, "F") ~ paste("FMT"),
                              str_detect(M_cat, "P") ~ paste("Placebo"),
                              str_detect(M_cat, "Other") ~ paste("Both"))) %>%
  bind_rows(data.frame(M_cat=c("F"), Species = c(0), Treatment="FMT")) %>%
  mutate(Treatment = factor(Treatment, levels = c("FMT", "Placebo", "Both"))) %>%
  ggplot(aes(Treatment, Species)) +
  geom_bar(stat = "identity") +
  scale_y_sqrt() +
  theme_bw() +
  theme(axis.title.x = element_blank()) +
  ylab("# of unique species engrafted in >= 6 patients")
ggsave("../figs/reads_CESp_6pts_bar.png",
       width = 8, height = 8, units = "cm")


```



## profile microbial taxa

```{r}

family_col= c(
"f__Lachnospiraceae" = "#6a3d9a",
"f__Ruminococcaceae" = "#33a02c",
"f__Bacteroidaceae" = "#1f78b4",
"f__Bacteroidales_unclassified" = "#ff7f00",
"f__Prevotellaceae" = "#b2df8a",
"f__Veillonellaceae" = "#fdbf6f",
"f__Bifidobacteriaceae" = "#ffff99",
"f__Clostridia_unclassified" =  "#a6cee3",
"f__Enterobacteriaceae" = "#e31a1c",
"f__Erysipelotrichaceae" = "#f781bf",
"f__Streptococcaceae" = "#cab2d6",
"f__Clostridiaceae" = "#fb9a99",
"f__Rikenellaceae" = "#b15928",
"f__Coriobacteriaceae" = "#8dd3c7",
"f__Bacilli_unclassified" = "#fb8072",
"f__Enterococcaceae" = "#fccde5",
"f__Desulfovibrionaceae" = "#80b1d3",
"f__Comamonadaceae" = "#b3de69",
"f__Peptostreptococcaceae" = "#ccebc5",
"f__Lactobacillaceae" = "#dfc27d",
"other" = "#808080"
)


```


```{r}

mapfile %>%
  mutate(sample= gsub(".R1.fastq.gz", "", File)) %>%
  mutate(Timepoint= case_when(grepl("DonorB_D_", sample) ~ paste("DMG"),
                              grepl("DonorB_P", sample) ~ paste("CEMG"),
                              TRUE ~ paste(Timepoint))) %>%
  mutate(Treatment= case_when(grepl("DonorB_D_", sample) ~ paste("DMG"),
                              grepl("DonorB_P", sample) ~ paste("CEMG"),
                              TRUE ~ paste(Treatment)))-> metaphlan_map
metaphlan %>%
  gather(sample, abundance, -clade_name) %>%
  left_join(metaphlan_map) %>%
  separate(clade_name, c("Kingdom", "Phylum", "Class", "Order", "Family", 
                         "Genus", "Species", "Strain"), sep = "[|]") %>%
  mutate(abundance= as.numeric(abundance)) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> tbl


################################
#select top 20 families
tbl %>%
  count(Family, wt = abundance) %>%
  filter(!is.na(Family)) %>%
  arrange(desc(n)) %>%
  pull(Family) -> tfam
  tfam20 <- tfam[1:20]
  
################################
# edit a new family column
tbl %>%
  mutate(Family_lab= case_when(Family %in% tfam20 ~ paste(Family),
                               TRUE ~ paste("other"))) -> tbl
# sum the abundance for each class, across all IDs, & sort the result
  tbl%>%
  count(Family_lab, wt = abundance) %>%
  arrange(desc(n)) %>%
  pull(Family_lab) -> sort.family
# get ID order, sorted by each ID's abundance in the most abundant class
  tbl %>%
  mutate(Fig_lab= case_when(is.na(Fig_lab) ~ paste(Study_ID),
                            TRUE ~ paste(Fig_lab))) %>%
  filter(Family_lab == sort.family[1]) %>%
  arrange(desc(abundance)) %>% distinct(Fig_lab) %>%
  pull(Fig_lab) -> ID.order

tbl$Fig_lab <- factor(tbl$Fig_lab, levels = ID.order)
tbl$Family_lab <- factor(tbl$Family_lab, levels = rev(sort.family))
tbl$Timepoint <- factor(tbl$Timepoint, levels = c("WK0", "WK6", "DMG", "CEMG"))
tbl$Treatment <- factor(tbl$Treatment, levels = c("Placebo", "FMT", "DMG", "CEMG"))


p1 <- tbl %>%
  filter(Treatment %in% c("Placebo", "FMT")) %>%
ggplot(aes(abundance, Fig_lab, fill=Family_lab, color=Family_lab)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = family_col) +
  scale_color_manual(values = family_col) +
  facet_grid(Treatment~Timepoint, scales = "free", space = "free") +
  theme_classic() +
  theme(legend.position = "none", 
        axis.title = element_blank())

p2 <- tbl %>%
  filter(Treatment %in% c("CEMG", "DMG")) %>%
ggplot(aes(abundance, sample, fill=Family_lab, color=Family_lab)) +
  geom_bar(stat = "identity") +
  facet_grid(Timepoint~., scales = "free", space = "free") +
  scale_fill_manual(values = family_col) +
  scale_color_manual(values = family_col) +
  theme_classic() +
  theme(legend.position = "left", 
        legend.title = element_blank(),
        axis.title = element_blank()) +
  guides(fill=guide_legend(ncol =1))

cowplot::plot_grid(p1, p2, ncol = 2)
ggsave("../figs/reads_RA_profile.png",
       width = 30, height = 15, units = "cm")

```

## Shannon diversity

```{r}

shannon <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                                "merged_metaphlan4_shannon.tsv", 
                      sep = ""), sep = "\t")

shannon %>%
  rownames_to_column("ID") %>%
  mutate(sample= gsub(".mp.profile_sp", "", ID)) %>%
  left_join(metaphlan_map) %>%
  # remove plate pools
  filter(!Timepoint %in% c("DMG","CEMG")) -> tbl

ggplot(tbl, aes(x = Timepoint, y = diversity_shannon)) +
  geom_boxplot(outlier.color = NA) +
  geom_line(aes(group=Fig_lab), position = position_dodge(0.2), alpha = 0.2) +
  #geom_jitter(aes(color = Timepoint), height = 0, width = .2) +
  geom_point(aes(fill=Remission,
                 group=Fig_lab),size=2,shape=21, 
             position = position_dodge(0.2)) + 
   facet_grid(~Treatment, scales = "free") +
   scale_fill_manual(values = c("#0571b0", "#ca0020"),
                       breaks = c("Res", "NoRes")) +
    ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Shannon")
ggsave("../figs/reads_Shannon_div.png",
       width = 6, height = 6, units = "cm")

# We happen to have lower diversity in patients on placebo; this observed both
# in baseline and post-FMT
# other than treatment group, the difference between timepoint is not sig diff
# in FMT compared to placebo
mod = lm(data=tbl, diversity_shannon  ~ Timepoint * Treatment)
anova(mod)
em = emmeans::emmeans(mod, "Timepoint", by = "Treatment")
as.data.frame(summary(em))

```

## Bray-curtis

```{r}

braycurtis <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                       "merged_metaphlan4_bray-curtis.tsv", 
                      sep = ""), sep = "\t")
colnames(braycurtis) <- gsub(".mp.profile_sp", "", colnames(braycurtis))
rownames(braycurtis) <- gsub(".mp.profile_sp", "", rownames(braycurtis))

braycurtis %>%
rownames_to_column("sample_1") %>%
  gather(sample_2, value, -sample_1) %>%
  filter(as.character(sample_1) != as.character(sample_2)) %>%
  mutate_if(is.factor, as.character) -> dis.tbl

# two mapfile for each variable in matrix
metaphlan_map -> V1.tbl
colnames(V1.tbl) <- paste(colnames(V1.tbl), "1", sep = "_")
metaphlan_map-> V2.tbl
colnames(V2.tbl) <- paste(colnames(V2.tbl), "2", sep = "_")

#merging all info together
dis.tbl %>%
  left_join(V1.tbl) %>%
  left_join(V2.tbl) %>%
  # remove plate pools
  filter(Treatment_1 != "CEMG" & Treatment_2 != "CEMG") %>%
  # compare samples within individuals
  filter(Fig_lab_1 == Fig_lab_2 | sample_1 == "DonorB_D_2013") %>%
  select(sample_1, sample_2, value, Fig_lab_1, Fig_lab_2,
         Timepoint_1, Timepoint_2, Treatment_1, Treatment_2,
         Remission_2) %>%
  mutate(Time_dis= case_when(Timepoint_1 == "DMG" & 
                               Timepoint_2 == "DMG" ~ paste("Donor"),
                             Timepoint_1 == "DMG" ~ paste("Donor", Timepoint_2,
                                                          sep = "-"),
                             TRUE ~ paste(Timepoint_1, Timepoint_2, 
                                          sep = "-"))) %>%
  # WE ARE SHOWING WK0<>WK6 which is same as below, so remove:
  filter(Time_dis != "WK6-WK0") %>%
  mutate(Remission_2= case_when(is.na(Remission_2) ~ paste("Donor"),
                                TRUE ~ paste(Remission_2)))-> tbl


ggplot(tbl, aes(Treatment_2, value)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Remission_2,
                 group=Fig_lab_1),size=2,shape=21, 
             position = position_dodge(0.2)) +
  facet_grid(~Time_dis, space = "free", scales = "free") +
  scale_fill_manual(values = c("#0571b0", "#ca0020", "#008837"),
                       breaks = c("Res", "NoRes", "Donor")) +
  ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Bray-Curtis")
ggsave("../figs/reads_Bray-Curtis_div.png",
       width = 10, height = 6, units = "cm")

```

## Aitchison

```{r}

aitchison <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                       "merged_metaphlan4_aitchison.tsv", 
                      sep = ""), sep = "\t")
colnames(aitchison) <- gsub(".mp.profile_sp", "", colnames(aitchison))
rownames(aitchison) <- gsub(".mp.profile_sp", "", rownames(aitchison))

aitchison %>%
rownames_to_column("sample_1") %>%
  gather(sample_2, value, -sample_1) %>%
  filter(as.character(sample_1) != as.character(sample_2)) %>%
  mutate_if(is.factor, as.character) -> dis.tbl

# two mapfile for each variable in matrix
metaphlan_map -> V1.tbl
colnames(V1.tbl) <- paste(colnames(V1.tbl), "1", sep = "_")
metaphlan_map-> V2.tbl
colnames(V2.tbl) <- paste(colnames(V2.tbl), "2", sep = "_")

#merging all info together
dis.tbl %>%
  left_join(V1.tbl) %>%
  left_join(V2.tbl) %>%
  # remove plate pools
  filter(Treatment_1 != "CEMG" & Treatment_2 != "CEMG") %>%
  # compare samples within individuals
  filter(Fig_lab_1 == Fig_lab_2 | sample_1 == "DonorB_D_2013") %>%
  select(sample_1, sample_2, value, Fig_lab_1, Fig_lab_2,
         Timepoint_1, Timepoint_2, Treatment_1, Treatment_2,
         Remission_2) %>%
  mutate(Time_dis= case_when(Timepoint_1 == "DMG" & 
                               Timepoint_2 == "DMG" ~ paste("Donor"),
                             Timepoint_1 == "DMG" ~ paste("Donor", Timepoint_2,
                                                          sep = "-"),
                             TRUE ~ paste(Timepoint_1, Timepoint_2, 
                                          sep = "-"))) %>%
  # WE ARE SHOWING WK0<>WK6 which is same as below, so remove:
  filter(Time_dis != "WK6-WK0") %>%
  mutate(Remission_2= case_when(is.na(Remission_2) ~ paste("Donor"),
                                TRUE ~ paste(Remission_2)))-> tbl

ggplot(tbl, aes(Treatment_2, value)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Remission_2,
                 group=Fig_lab_1),size=2,shape=21, 
             position = position_dodge(0.2)) +
  facet_grid(~Time_dis, space = "free", scales = "free") +
  scale_fill_manual(values = c("#0571b0", "#ca0020", "#008837"),
                       breaks = c("Res", "NoRes", "Donor")) +
  ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Aitchison")
ggsave("../figs/reads_Aitchison_div.png",
       width = 10, height = 6, units = "cm")


```



