---
title: "Responder gene signature"
author: "Sharok"
date: "2023-10-02"
output: html_document
---


```{r}

library(tidyverse)
library(tidytext)

```


```{r}

mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

```

## contigs in bins
```{r}

# map information for contigs and bins
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", header = F)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", header = F)
colnames(c_inBin) <- c("contig", "bin_id")
colnames(c_inMAG) <- c("contig", "bin_id")

c_inMAG %>% select(bin_id) %>% distinct() %>% pull() -> MAGs


```

## bins taxonomy

```{r}

# gtdb taxonomy output for all the bins
gtdb.bac <- read.csv("../data/Assembly/gtdbtk.bac120.summary.tsv", sep = "\t")
gtdb.arc <- read.csv("../data/Assembly/gtdbtk.ar122.summary.tsv", sep = "\t")

gtdb.bac %>% mutate(bin_id=user_genome) %>%
  select(bin_id, classification) %>%
  bind_rows(gtdb.arc %>% mutate(bin_id=user_genome) %>%
              select(bin_id, classification)) %>%
  mutate(bin_id=gsub("bin.", "bin_", bin_id)) -> gtdb
  
```

## Genes and Contigs in gff

```{r}

gff_cols <- c("seqname","source","feature","start","end","score","strand",
              "frame","attributes")
read.delim("../data/perfect_mapping/B_db_1kb.gff3",
           header=F, comment.char="#", ) -> gff
colnames(gff) <- gff_cols

# removing sequence info from the end of the gff file
gff %>%
  filter(row_number() < 1224073) %>%
  mutate(ID= gsub(";.*", "", attributes)) %>%
  mutate(ID=gsub("ID=", "", ID)) %>%
  # annotated regions that have only contig info in attributes. 
  # I don't need them
  filter(feature!= "region")-> gff

```

## Genes commonly engrafted in >= 3 responder

```{r}

int_genes <- read.csv("../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res.txt",
                      header = T, sep = "\t")
int_genes %>% rename(ID=Marker) %>% filter(n > 0) -> int_genes

```


```{r}

int_genes %>%
  left_join(gff) %>%
  left_join(c_inBin %>% mutate(seqname=contig)) %>%
  mutate(bin_id= case_when(is.na(bin_id) ~ paste("NoBin"),
                           TRUE ~ paste(bin_id))) -> int_genes_inBins


int_genes_inBins %>%
  group_by(Fig_lab, bin_id) %>%
  tally() %>%
  mutate(Type= case_when(bin_id %in% MAGs ~ paste("MAG"),
                         bin_id == "NoBin" ~ paste("NoBin"),
                         TRUE ~ paste("Bin"))) %>%
 mutate(gene_num= case_when(n <= 10 ~ paste("<=10"),
                             between(n, 10, 50) ~ paste("10-50"),
                             between(n, 50, 250) ~ paste("50-250"),
                             between(n, 250, 550) ~ paste("250-550"),
                             TRUE ~ paste("other"))) -> plt_df



ggplot(plt_df, aes(reorder_within(bin_id, -n, Type), 
                   reorder_within(Fig_lab, -n, 1), fill=gene_num)) +
  geom_tile() +
  facet_grid(~Type, space = "free", scales = "free") +
   scale_fill_manual(values = c("<=10" = "#bae4bc",
                               "10-50" = "#7bccc4",
                               "50-250" = "#43a2ca",
                               "250-550" = "#0868ac"), na.value = "white") +
  theme_classic() +
  scale_x_reordered() + scale_y_reordered() +
  theme(axis.text.x = element_blank(),
        axis.title = element_blank(),
        legend.position = "none",
        strip.text.x = element_blank())
ggsave("../figs/genes_CEm_3pts_inBins.png",
       width = 12, height = 3, units = "cm")


int_genes_inBins %>%
  mutate(Type= case_when(bin_id %in% MAGs ~ paste("MAG"),
                         bin_id == "NoBin" ~ paste("NoBin"),
                         TRUE ~ paste("Bin"))) %>%
  select(ID, bin_id, Type) %>% distinct() -> int_genes_inBins.stat

int_genes_inBins.stat %>%
  group_by(Type) %>%
  summarize(genes=n())

int_genes_inBins.stat %>%
  group_by(Type, bin_id) %>%
  tally() %>%
  group_by(Type) %>%
  summarize(bins=n())


```


## saving gene list
for further investigation in RefSeq and a SRA dataset

```{r}

int_genes %>%
  select(ID) %>% distinct() %>%
  write.table("../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList.txt",
            sep = "\t", quote = F, col.names = F, row.names = F)

```



## commonly engrafted genes in RefSeq:
this is the result of blastn. a local refseq db was built on May 23rd, 2023

```{r}

blastn_parser <- function(table1, table2, identity, coverage) {
  # table1: .fna.fai file of genes, showing genes length, start and stop
  # table2: .blastout standard blastout of genes.fna against refseq
  # identity: pident cutoff
  # coverage: query coverage cutoff

gene_length <- read.csv(table1, sep = "\t", header = F)
gene_length %>%
  select(V1, V2) %>%
  rename(qseqid=V1,
         gene_length=V2) -> gene_length

blastout <- read.csv(table2, sep = "\t", header = F)
col_names <- c("qseqid", "sseqid", "pident", "length",
              "mismatch", "gapopen", "qstart", 
              "qend", "sstart", "send",
              "evalue", "bitscore")

colnames(blastout) <- col_names
blastout %>%
  left_join(gene_length) %>%
  # gene length must be over 100bp
  # pident must be over 95
  filter(gene_length >= 100 & pident >= identity) %>%
  group_by(qseqid, sseqid, gene_length) %>%
  summarise(total_length= sum(length),
            hit_count=n()) %>%
  ungroup() %>%
  mutate(query_coverage= total_length/gene_length*100) %>%
  # over 95% of gene of interest must be covered
  filter(query_coverage >= coverage)

}

```



```{r}

refseq_name <- read.csv("../data/genes_of_interest/references_headers.txt",
                        sep = "\t", header = F)
refseq_name %>%
  mutate(sseqid= gsub(" .*", "", V1)) %>%
  mutate(sseqid= gsub(">", "", sseqid)) %>%
  mutate(filename=gsub(".*__", "", V1)) %>%
  rename(refseq_header=V1)-> refseq_name
gtdb_refseq <- read.csv("../data/genes_of_interest/gtdb.bac120.summary.tsv",
                        sep = "\t") 
gtdb_refseq %>%
  select(user_genome, classification) %>%
  separate(classification, c("Kingdom", "Phylum", "Class",
                                   "Order", "Family", "Genus",
                                   "Species"), ";.__") %>%
  mutate(filename= paste(user_genome, ".fna.gz", sep = "")) %>%
  left_join(refseq_name) -> refseq_taxonomy

################################################################################

table1 <- "../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList.fna.fai"
table2 <- "../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList_inRefSeq.blastout"
blastn_parser(table1, table2, identity = 95, coverage = 95) %>%
  group_by(sseqid) %>%
  summarise(genes=n()) %>%
  left_join(refseq_taxonomy) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(spID= paste(Species, row_number(), sep = "_")) -> refseq_in3pts
# genomes with >= 5 genes:
refseq_in3pts %>%
  filter(genes >= 5) %>%
ggplot(aes(genes, reorder_within(spID, genes, Family), fill="CEGs")) +
  geom_bar(stat = "identity") +
  facet_grid(Family~., space = "free", scales = "free") +
  scale_x_log10() +
  scale_fill_manual(values = c("CEGs" = "#008837")) +
  scale_y_reordered() +
  theme_classic() +
  theme(strip.text.y = element_blank(),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title = element_blank())
ggsave("../figs/genesInRefSeq_3pts_bar.png",
       width = 6, height = 12, units = "cm")


refseq_in3pts %>%
  filter(genes >= 5) -> test


```



## comparison of genes across genomes

```{r}

refseq_in3pts %>%
  ungroup %>%
  filter(genes >= 5) %>%
  select(sseqid) %>% distinct() %>% pull() -> top_genome_in3pts
table1 <- "../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList.fna.fai"
table2 <- "../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList_inRefSeq.blastout"
blastn_parser(table1, table2, identity = 95, coverage = 95) %>%
  filter(sseqid %in% top_genome_in3pts) %>%
  group_by(sseqid, qseqid) %>%
  summarise(genes=n()) %>%
  left_join(refseq_taxonomy) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(spID= paste(Species, row_number(), sep = "_")) %>%
# visualize
ggplot(aes(qseqid, reorder_within(sseqid, genes, Species), fill="CEGs")) +
  geom_tile() +
  facet_grid(Family~Genus, space = "free", scales = "free") +
  scale_fill_manual(values = c("CEGs" = "#008837")) +
  scale_y_reordered() +
  theme_classic() +
  theme(strip.text.y = element_blank(),
        strip.text.x = element_blank(),
        legend.position = "none",
        axis.text = element_blank(),
        axis.title = element_blank())
ggsave("../figs/genesInRefSeq_3pts_heatmap.png",
       width = 6, height = 12, units = "cm")

```



```{r}

refseq_in3pts %>%
  ungroup %>%
  filter(genes >= 5) -> top_refseq_in3pts
top_refseq_in3pts %>%
  select(Genus) %>% distinct() %>% pull() -> refseq_in3pts_genus

refseq_taxonomy %>%
  filter(Genus %in% refseq_in3pts_genus) %>%
  left_join(top_refseq_in3pts) %>%
  mutate(genes = case_when(is.na(genes) ~ paste(0),
                           TRUE ~ paste(genes))) %>%
  group_by(Species) %>%
  mutate(spID2= paste(Species, row_number(), sep = "_")) %>%
  ggplot(aes(as.numeric(genes), reorder_within(spID2, as.numeric(genes), 
                                               Genus), fill="CEGs")) +
  geom_bar(stat = "identity") +
  facet_grid(Genus~.,space = "free", scales = "free") +
  scale_fill_manual(values = c("CEGs" = "#008837")) +
  scale_y_reordered() +
  theme_classic() +
  theme(strip.text.y = element_text(angle = 0),
        legend.position = "none",
        axis.text.y = element_blank(),
        axis.title = element_blank())

```


## Functions of commonly engrafted genes in respoders
proteins were clusterd at 90% using mmseq and eggnog was used for functional
annotations of the proteins


## KEGG parser
a list of function to parse KEGGs from eggnot outputs
```{r}

library(KEGGREST)

###################################################
# function to parse eggnog output fot KEGG pathway:
egg_KEGGpathway <- function(file.path) {
  
  # get all the list of all pathways:
  data.frame(keggList(database = "PATHWAY")) -> KEGG.PATHWAY
  colnames(KEGG.PATHWAY) <- c("name")
  KEGG.PATHWAY %>%
  rownames_to_column("KEGG_Pathway") -> KEGG.PATHWAY
  
  eggnog <- read.csv(file.path, sep = "\t")
  eggnog %>%
  select(KEGG_Pathway) %>% filter(!grepl("-", KEGG_Pathway)) %>%
  separate_rows(1,sep = ",") %>% group_by(KEGG_Pathway) %>% tally() %>%
  # only pathways
  filter(grepl("map", KEGG_Pathway)) %>%
  left_join(KEGG.PATHWAY) %>%
  group_by(name) %>%
  summarise(n=sum(n))
}

###################################################
# function to parse eggnog output for KEGG ko:
egg_KEGGko <- function(file.path) {
  
  # get the list of all ko:
  data.frame(keggList(database = "KO")) -> KEGG.KO
  colnames(KEGG.KO) <- c("name")
  KEGG.KO %>%
  rownames_to_column("KEGG_ko") -> KEGG.KO
  
  eggnog <- read.csv(file.path, sep = "\t")
  eggnog %>%
  select(KEGG_ko) %>%
  filter(!grepl("-", KEGG_ko)) %>%
  separate_rows(1,sep = ",") %>% group_by(KEGG_ko) %>% tally() %>%
  mutate(KEGG_ko= gsub("ko:", "", KEGG_ko)) %>%
  left_join(KEGG.KO) %>%
  group_by(name) %>%
  summarise(n=sum(n))
}

###################################################
# function to parse eggnog output for KEGG Module:
egg_KEGGmodule <- function(file.path) {
  
  # get the list of all MODULE:
  data.frame(keggList(database = "MODULE")) -> KEGG.MODULE
  colnames(KEGG.MODULE) <- c("name")
  KEGG.MODULE %>%
  rownames_to_column("KEGG_Module") -> KEGG.MODULE
  
  eggnog <- read.csv(file.path, sep = "\t")
  eggnog %>%
  select(KEGG_Module) %>%
  filter(!grepl("-", KEGG_Module)) %>%
  separate_rows(1,sep = ",") %>% group_by(KEGG_Module) %>% tally() %>%
  left_join(KEGG.MODULE) %>%
  group_by(name) %>%
  summarise(n=sum(n))
}

###################################################
# function to parse eggnog output for KEGG BRITE:
egg_KEGGbrite <- function(file.path) {
  
  # get the list of all BRITE:
  data.frame(keggList(database = "BRITE")) -> KEGG.BRITE
  colnames(KEGG.BRITE) <- c("name")
  KEGG.BRITE %>%
  rownames_to_column("BRITE") -> KEGG.BRITE
  
  eggnog <- read.csv(file.path, sep = "\t")
  eggnog %>%
  select(BRITE) %>%
  filter(!grepl("-", BRITE)) %>%
  separate_rows(1,sep = ",") %>% group_by(BRITE) %>% tally() %>%
  left_join(KEGG.BRITE) %>%
  group_by(name) %>%
  summarise(n=sum(n))
}

###################################################
# function to parse eggnog output for KEGG Reaction:
egg_KEGGreaction <- function(file.path) {
  
  # get the list of all REACTION:
  data.frame(keggList(database = "REACTION")) -> KEGG.REACTION
  colnames(KEGG.REACTION) <- c("name")
  KEGG.REACTION %>%
  rownames_to_column("KEGG_Reaction") -> KEGG.REACTION
  
  eggnog <- read.csv(file.path, sep = "\t")
  eggnog %>%
  select(KEGG_Reaction) %>%
  filter(!grepl("-", KEGG_Reaction)) %>%
  separate_rows(1,sep = ",") %>% group_by(KEGG_Reaction) %>% tally() %>%
  left_join(KEGG.REACTION) %>%
  group_by(name) %>%
  summarise(n=sum(n))
}

##################################################
# A function to parse all KEGG categories together
egg_KEGGwraper <- function(file.path) {
  
  egg_KEGGbrite(file.path) %>%
    mutate(Category=paste("BRITE")) -> tb1
  egg_KEGGpathway(file.path) %>%
    mutate(Category=paste("PATHWAY")) -> tb2
  egg_KEGGmodule(file.path) %>%
    mutate(Category=paste("MODULE")) -> tb3
  egg_KEGGko(file.path) %>%
    mutate(Category=paste("KO")) -> tb4
  egg_KEGGreaction(file.path) %>%
    mutate(Category=paste("REACTION")) -> tb5
  rbind(tb1, tb2, tb3, tb4, tb5) %>%
    arrange(Category, -n)
}


```

## KEGG pathway 
using eggnog data: reading eggnog outputs
```{r}


# Commonly engrafted genes in >= 3 pts only responder patients:
input <- "../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList_90c.emapper.annotations"
eggnog <- read.csv(input, sep = "\t")
egg_KEGGwraper(input) -> tbl
#write.table(tbl, output, sep = "\t", row.names = F, quote = F)

eggnog %>%
  select(EC) %>%
  mutate(type=case_when(EC == "-" ~ paste("unknown EC"),
                        TRUE ~ paste("known EC"))) %>%
  group_by(type) %>% tally()

eggnog %>%
  select(KEGG_ko) %>%
  mutate(type=case_when(KEGG_ko == "-" ~ paste("unknown KEGG_ko"),
                        TRUE ~ paste("known KEGG_ko"))) %>%
  group_by(type) %>% tally()

tbl %>%
  filter(Category == "BRITE") %>%
  mutate(total_n= sum(n)) %>%
  mutate(percent= n/sum(n) * 100) -> test

tbl %>%
  filter(Category == "PATHWAY") %>%
  mutate(total_n= sum(n)) %>%
  mutate(percent= n/sum(n) * 100) %>%
  mutate(class= case_when(is.na(name) ~ paste("Unknown"),
                          percent >= 2 ~ paste("top2%"),
                          TRUE ~ paste("Other")
                          )) -> tbl2

color1 <- c("Other"= "#b2df8a",
            "top2%" = "#33a02c",
            "Unknown"= "#bababa")
tbl2 %>%
  group_by(class) %>%
  summarise(n= sum(n)) %>%
  mutate(percent= n/sum(n)* 100) %>%
  mutate(Category= paste("PATHWAY")) %>%
  mutate(class= factor(class, levels = c("top2%","Other", "Unknown"))) %>%
ggplot(aes(Category, percent, fill=class)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = round(percent, 1)),
            position = position_stack(vjust = .5), angle=0) +
  scale_fill_manual(values = color1) +
  theme_classic() +
  theme(axis.title = element_blank(),
        legend.position = "right",
        legend.title = element_blank())
ggsave("../figs/CEGs_Func1.png",
       width = 6, height = 5, units = "cm")

tbl2 %>%
  filter(class== "top2%") %>%
  ggplot(aes(percent, reorder_within(name, percent, class), fill=class)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = color1) +
  scale_y_reordered() +
  theme_classic() +
  theme(axis.title = element_blank(),
        legend.position = "none")
ggsave("../figs/CEGs_Func2.png",
       width = 10, height = 5, units = "cm")

```


## Metagenomic survey for CEGs in responders
Reading mapfiles for SRA studies:

```{r, echo=FALSE, message=FALSE}

mpath="../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res_geneList_PRJNA400072/"
Run <- read.csv(paste(mpath, "PRJNA400072_SRA.txt", sep = ""))
Diag <- read.csv(paste(mpath, "PRJNA400072_Diagnosis.txt", sep = ""))
Diag %>%
  mutate(Sample.Name= case_when(grepl("^[0-9]", 
                                      local_sample_id) ~ paste("PRISM",
                                                               local_sample_id,
                                                               sep = "_"),
                                TRUE ~ as.character(local_sample_id))) %>%
  select(Sample.Name, Diagnosis) %>%
  left_join(Run, by = "Sample.Name") -> PRJNA400072

############################################################################

#total seqeuncing depth for each sample:
SRAreads <- read.csv(paste(mpath, "sequencing_depth.txt", sep = ""),
                     sep = "\t")
PRJNA400072 %>%
  select(Run, Sample.Name, Diagnosis) %>%
  rename(SampleName=Sample.Name) %>%
  mutate(Dataset= paste("PRJNA400072")) %>%
  mutate(sample.id=Run) %>%
  left_join(SRAreads)-> SRAmapfile

table(SRAmapfile$Diagnosis)

```



```{r}
cols <- c("sample.id", "rname", "startpos", "endpos", 
          "numreads","covbases",
          "coverage", "meandepth", "meanbaseq", "meanmapq")

path=paste(mpath, "PRJNA400072", sep = "")
patt=".cover"
data.frame(sample.id = paste(dir(path, 
                                 pattern = patt))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read_tsv(
  file.path(path, .), 
            col_names = F))) %>%
         unnest() %>%
  mutate(sample.id = gsub(patt, "", sample.id)) %>%
  filter(X2 != "startpos") -> SRAcoverage
colnames(SRAcoverage) <- cols


SRAcoverage %>%
  mutate(cover_type= case_when(as.numeric(coverage) >= 100 ~ paste("Pos"),
                               TRUE ~ paste("Neg"))) %>%
  group_by(sample.id, cover_type) %>% tally() -> SRA_cegs_counts



SRAcoverage %>%
  mutate(cover_type= case_when(as.numeric(coverage) >= 100 ~ paste("Pos"),
                               TRUE ~ paste("Neg"))) %>%
  group_by(sample.id, cover_type) %>%
  summarise(reads= sum(as.numeric(numreads))) %>%
  left_join(SRAreads) %>%
  mutate(relativeAbund= 
           as.numeric(reads) / as.numeric(Total_reads) * 100)-> SRA_cegs_abund

```


```{r}

SRAmapfile %>%
  left_join(SRA_cegs_counts) %>%
  filter(cover_type== "Pos") %>%
  mutate(Diagnosis= factor(Diagnosis, 
                           levels = c("Control", "UC", "CD"))) -> tbl1

model=lm( tbl1$n ~ tbl1$Diagnosis )
ANOVA=aov(model)
TUKEY <- TukeyHSD(x=ANOVA, 'tbl1$Diagnosis', conf.level=0.95)
TUKEY
plot(TUKEY , las=1 , col="brown")

ggplot(tbl1, aes(Diagnosis, n)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Diagnosis),size=2,shape=21, 
             position = position_dodge(0.2)) + 
  scale_fill_manual(values = c("Control" = "#1a9641",
                                "UC" = "#d7191c",
                                "CD" = "#d7191c")) +
  theme_classic() +
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")
  #ylab("# of unique CEGs / sample")
ggsave("../figs/CEGs_inPublicMeta_unique.png",
       width = 5, height = 5, units = "cm")

############################################################################

SRAmapfile %>%
  left_join(SRA_cegs_abund) %>%
  filter(cover_type== "Pos") %>%
  mutate(Diagnosis= factor(Diagnosis, 
                           levels = c("Control", "UC", "CD"))) -> tbl2

model=lm( tbl2$relativeAbund ~ tbl2$Diagnosis )
ANOVA=aov(model)
TUKEY <- TukeyHSD(x=ANOVA, 'tbl2$Diagnosis', conf.level=0.95)
TUKEY
plot(TUKEY , las=1 , col="brown")

ggplot(tbl2, aes(Diagnosis, relativeAbund)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Diagnosis),size=2,shape=21, 
             position = position_dodge(0.2)) + 
  scale_fill_manual(values = c("Control" = "#1a9641",
                                "UC" = "#d7191c",
                                "CD" = "#d7191c")) +
  theme_classic() +
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")
  #ylab("% of CEGs reads / sample")
ggsave("../figs/CEGs_inPublicMeta_abundance.png",
       width = 5, height = 5, units = "cm")


```

