---
title: "MAGs analysis"
author: "Sharok"
date: "5/4/2023"
output: html_document
---

```{r}
library(tidyverse)
library(tidytext)
```


```{r}
mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

```

## contigs in bins
```{r}

# map information for contigs and bins
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", header = F)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", header = F)
colnames(c_inBin) <- c("contig", "bin_id")
colnames(c_inMAG) <- c("contig", "bin_id")

c_inMAG %>% select(bin_id) %>% distinct() %>% pull() -> MAGs

```

## bins taxonomy
```{r}

# gtdb taxonomy output for all the bins

gtdb.bac <- read.csv("../data/Assembly/gtdbtk.bac120.summary.tsv", sep = "\t")
gtdb.arc <- read.csv("../data/Assembly/gtdbtk.ar122.summary.tsv", sep = "\t")

gtdb.bac %>% mutate(bin_id=user_genome) %>%
  select(bin_id, classification) %>%
  bind_rows(gtdb.arc %>% mutate(bin_id=user_genome) %>%
              select(bin_id, classification)) %>%
  mutate(bin_id=gsub("bin.", "bin_", bin_id)) -> gtdb
  
```

## Perfect alignment 
Mapping patients and donor B reads to  mapped to donB and UHGG databases using
bwa-mem (see cutoff)

```{r}

# changing names so that it works with visualizing functions below:
tmp.names <- c("sample.id", "bin_id", "bin_len", "bin_covbases",
               "bin_numreads", "bin_coverage")

UHGG_perfect_1x <- read.csv(paste("../data/perfect_mapping/",
                                  "UHGG/UHGG_1x_cover_perfect.csv",
                                  sep = ""))
colnames(UHGG_perfect_1x) <- tmp.names

# ONLY patiens:
UHGG_perfect_1x %>%
  mutate(sample.id= gsub("_perfect_UHGG.cover", "", sample.id)) %>%
  filter(str_detect(sample.id, "PMCL")) %>%
  dplyr::rename(sample=sample.id) -> UHGG_perfect_1x_P

# ONLY donor
UHGG_perfect_1x %>%
  mutate(sample.id= gsub("_perfect_UHGG.cover", "", sample.id)) %>%
  filter(str_detect(sample.id, "DonorB")) %>%
  mutate(sample.id= gsub("DonorB_D_", "", sample.id)) %>%
  dplyr::rename(sample=sample.id) -> UHGG_perfect_1x_D

############################################################################
# Perfect alignment of Donor Breads to donB database via bwa-mem (see cutoff)
############################################################################

heads_cover <- c("sample.id",
                 "contig", "startpos", "endpos", 
                 "numreads", "covbases", "coverage", "meandepth",
                 "meanbaseq", "meanmapq",
                 "ReadSampling", "Quality")

path="../data/perfect_mapping/DBD"
patt="_perfect_.*.cover"
data.frame(sample.id = paste(dir(path,
                                 pattern = patt))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read.csv(file.path(path, .),
                                                 sep = "\t", header = F,
                                                 comment.char = "#"))) %>%
unnest() %>% 
mutate(sample.id = gsub("_perfect_BDB.cover", "", sample.id)) -> BDB_perfect
colnames(BDB_perfect) <- heads_cover

BDB_perfect %>%
  left_join(c_inBin) %>%
  mutate(bin_q= case_when(bin_id %in% MAGs ~ paste("MAG"),
                          TRUE ~ paste("Bin"))) %>%
  filter(bin_q == "MAG")  %>% select(-bin_q) %>%
  dplyr::rename(sample=sample.id) %>%
  group_by(sample, bin_id) %>%
  summarise(bin_len= sum(endpos),
            bin_covbases=sum(covbases),
            bin_numreads=sum(numreads),
            bin_coverage=bin_covbases/bin_len*100) -> BDB_perfect_1x

BDB_perfect_1x %>%
  filter(str_detect(sample, "PMCL")) -> BDB_perfect_1x_P

BDB_perfect_1x %>%
  filter(str_detect(sample, "DonorB")) %>%
  mutate(sample= gsub("DonorB_D_", "", sample))  -> BDB_perfect_1x_D

##############################################################################

# UHGG + BDB together for donor sample:
UHGG_perfect_1x_D %>%
  bind_rows(BDB_perfect_1x_D) %>%
  mutate(DB= case_when(str_detect(bin_id, "UHGG") ~ paste("UHGG"),
                       str_detect(bin_id, "bin") ~ paste("BDB"),
                       TRUE ~ paste("Other"))) -> table_D
# UHGG + BDB together for patient sample:
UHGG_perfect_1x_P %>%
  bind_rows(BDB_perfect_1x_P) %>%
  mutate(DB= case_when(str_detect(bin_id, "UHGG") ~ paste("UHGG"),
                       str_detect(bin_id, "bin") ~ paste("BDB"),
                       TRUE ~ paste("Other"))) -> table_P

 
# calculate percentage of positive bins per sample
Fpos <- function(table, cutoff){
  
  table %>%
  mutate(detect= case_when(bin_coverage >= cutoff ~ paste("Yes"),
                           TRUE ~ paste("No"))) %>%
  filter(detect == "Yes") %>%
  group_by(sample, DB) %>%
  summarise(pos_n=n(),
            total=length(unique(table$bin_id)),
            pos_p=pos_n/length(unique(table$bin_id))*100) %>%
  mutate(cutoff=paste(cutoff))
}

# comparison of detection cut-offs
Fpos_var <- function(table) {
  
vector <- seq(10, 100, by=10)
res <- list()
for (i in vector) {
   res[[i]] <- Fpos(table = table, i)
}
do.call(rbind.data.frame, res)
}

```

## benchmarking- donor
visualizing only donor samples, comparing detectin cutoffs

```{r}

Fpos_var(table_D) -> tbl

tbl %>%
ungroup() %>%
mutate(sample= gsub("sub_", "", sample))-> tbl

cols <- c("2013" = "#bae4b3",
          "2016" = "#74c476", 
          "May17" = "#31a354",
          "Oct17" = "#006d2c")

vector <- seq(10, 100, by=10)
tbl$cutoff <- factor(tbl$cutoff, levels = vector)
ggplot(tbl, aes(as.numeric(as.character(cutoff)), pos_n, color=sample)) +
  facet_grid(~DB) +
  geom_line(aes(group=sample), size = 1) +
  geom_vline(xintercept = 20, linetype="dotted", 
                color = "black", size=0.5, alpha=0.3) +
  annotate("text", x=16, y=200, label="20%", angle=90, alpha=0.3) +
  geom_vline(xintercept = 90, linetype="dashed", 
                color = "black", size=0.5, alpha = 0.3) +
  annotate("text", x=86, y=200, label="90%", angle=90, alpha=0.3) +
  scale_color_manual(values = cols) +
  theme_classic() +
  theme(legend.position = c(x=0.2, y=0.8),
        legend.title = element_blank()) +
  xlab("Detection cutoff") +
  ylab("# of detected bins")
ggsave("../figs/Benchmark_perfect_BDB_UHGG.png",
       width = 16, height = 10, units = "cm")


```

## benchmarking- pts
visualizing only patients samples, comparing detectin cutoffs

```{r}

Fpos_var(table_P) -> tbl

tbl %>%
  ungroup() %>%
  mutate(sample= gsub("sub_", "", sample)) %>%
  mutate(Study_ID= gsub("_.*", "", sample)) %>%
  left_join(mapfile %>% select(Study_ID, Treatment, Timepoint, Remission))-> tbl

cols <- c("NoRes" = "#d7191c",
          "Res" = "#2b83ba")

tbl$cutoff <- factor(tbl$cutoff, levels = vector)
ggplot(tbl, aes(as.numeric(as.character(cutoff)), pos_n, color=Remission)) +
  facet_grid(DB~Treatment+Timepoint) +
  geom_line(aes(group=sample), size = 1) +
  geom_vline(xintercept = 20, linetype="dotted", 
                color = "black", size=0.5, alpha=0.3) +
  annotate("text", x=16, y=200, label="20%", angle=90, alpha=0.3) +
  geom_vline(xintercept = 90, linetype="dashed", 
                color = "black", size=0.5, alpha = 0.3) +
  annotate("text", x=86, y=200, label="90%", angle=90, alpha=0.3) +
  scale_color_manual(values = cols) +
  scale_y_log10() +
  theme_classic() +
  theme(legend.position = "bottom",
        legend.title = element_blank()) +
  xlab("Detection cutoff") +
  ylab("# of detected bins")
ggsave("../figs/Benchmark_perfect_BDB_UHGG_patient.png",
       width = 16, height = 10, units = "cm")


```

## Engrafted MAGs

```{r}

C=90
table_P %>%
  ungroup() %>%
  mutate(sample= gsub("sub_", "", sample)) %>%
  mutate(Study_ID= gsub("_.*", "", sample)) %>%
  left_join(mapfile %>% select(Study_ID, Fig_lab, Timepoint)) %>%
  select(Fig_lab, bin_id, bin_coverage, DB, Timepoint) %>%
  spread(Timepoint, bin_coverage) %>%
  mutate(detection= case_when(WK0 >= C & WK6 >= C ~ paste("Both"),
                              WK0 >= C & WK6 < C ~ paste("Pre"),
                              WK0 < C & WK6 >= C ~ paste("Post"),
                              WK0 < C & WK6 < C ~ paste("none"),
                              TRUE ~ paste("Other"))) -> tbl

tbl %>%
  group_by(Fig_lab, DB, detection) %>%
  summarise(n=n()) %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> tbl0


cols <- c("none" = "#bababa",
          "Pre" = "#7b3294",
          "Both" = "#a6dba0",
          "Post" = "#008837")

###############################################################################
#1. summary per treatment:
tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment) %>%
            distinct()) %>%
  group_by(DB, Treatment) %>%
  tally(name = "total") -> total_treat

tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  group_by(DB, detection, Treatment) %>%
  summarise(n=n()) %>%
  left_join(total_treat) %>%
  mutate(percent=n/total*100) -> tbl1

tbl1 %>%
mutate(detection= factor(detection, 
                        levels = c("Post","Both","Pre", "none"))) %>%
# only BDB database
filter(DB=="BDB") %>%
# don't show percent valus less than 1
ggplot(aes(Treatment, percent,fill=detection)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = round(percent, 1)),
            position = position_stack(vjust = .5), angle=0,
            size=2) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "right", legend.title = element_blank(),
        axis.title = element_blank())
ggsave("../figs/MAGs_perfect_treatment.png",
       width = 6, height = 8, units = "cm")
################################################################################
###############################################################################
#2. summary per patients:
tbl0 %>%
  filter(DB=="BDB") %>%
  mutate(detection= factor(detection, 
                           levels = c("none", "Pre", "Both", "Post"))) %>%
  ggplot(aes(n, Fig_lab, fill=detection)) +
  geom_bar(stat = "identity") +
  facet_grid(Treatment~., space = "free", scales = "free") +
  scale_x_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none", axis.title = element_blank())
ggsave("../figs/MAGs_perfect_BDB_patient.png",
       width = 8, height = 8, units = "cm")
################################################################################
###############################################################################
#3. engraftment stats:
# some of pts dont have any engrafted mags, those should have
# 0 engraftment to properly give weight to them 
tbl0 %>%
  select(DB, Fig_lab, Treatment, Remission) %>%
  distinct() %>%
  left_join(tbl0 %>% filter(detection == "Post")) %>%
  mutate(n= case_when(is.na(n) ~ paste(0),
                      TRUE ~ paste(n))) %>%
  mutate(detection= case_when(is.na(detection) ~ paste("Post"),
                      TRUE ~ paste(detection))) %>%
  #only BDB database
  filter(DB=="BDB")-> eng_mags

 ggplot(eng_mags, aes(Remission, as.numeric(n), color= detection)) +
  geom_boxplot(outlier.color = NA) +
  geom_jitter() +
  facet_grid(~Treatment, scales = "free") +
  scale_color_manual(values = cols) +
  scale_y_sqrt() +
  theme_classic() +
  theme(legend.position = "none", axis.title = element_blank())
ggsave("../figs/MAGs_perfect_engraftment.png",
       width = 6, height = 8, units = "cm")
```

## Taxonomy of engrafted MAGs

```{r}


C=90
table_P %>%
  ungroup() %>%
  mutate(sample= gsub("sub_", "", sample)) %>%
  mutate(Study_ID= gsub("_.*", "", sample)) %>%
  left_join(mapfile %>% select(Study_ID, Fig_lab, Timepoint)) %>%
  select(Fig_lab, bin_id, bin_coverage, DB, Timepoint) %>%
  spread(Timepoint, bin_coverage) %>%
  mutate(detection= case_when(WK0 >= C & WK6 >= C ~ paste("Both"),
                              WK0 >= C & WK6 < C ~ paste("Pre"),
                              WK0 < C & WK6 >= C ~ paste("Post"),
                              WK0 < C & WK6 < C ~ paste("none"),
                              TRUE ~ paste("Other"))) -> tbl


gtdb %>%
  separate(classification, c("Kingdom", "Phylum", "Class",
                                   "Order", "Family", "Genus",
                                   "Species"), ";.__") -> gtdb_viz
#select top 10 families
gtdb_viz %>%
  filter(bin_id %in% MAGs) %>%
  group_by(Family) %>%
  summarise(n= n()) %>%
  arrange(-n) %>% pull(Family) -> tfam
tfam10 <- tfam[1:10]

##############################################################################
## 1. summary of taxonomy of donor B MAGs
##############################################################################
family_col= c(
"Lachnospiraceae" = "#6a3d9a",
"Ruminococcaceae" = "#33a02c",
"Bacteroidaceae" = "#1f78b4",
"Oscillospiraceae" = "#ff7f00",
"CAG-508" = "#b2df8a",
"Acutalibacteraceae" = "#fdbf6f",
"Streptococcaceae" = "#ffff99",
"CAG-822" =  "#a6cee3",
"Lactobacillaceae" = "#fb9a99",
"Butyricicoccaceae" = "#f781bf",
"Other" = "#808080"
)

family_order <- c("Lachnospiraceae", "Oscillospiraceae",
                  "Ruminococcaceae", "Bacteroidaceae", "CAG-508",
                  "Acutalibacteraceae", "Streptococcaceae",
                  "Lactobacillaceae", "Butyricicoccaceae",
                  "CAG-822", "Other")

gtdb_viz %>%
  filter(bin_id %in% MAGs) %>%
  mutate(Family_lab= case_when(Family %in% tfam10 ~ paste(Family),
                               TRUE ~ paste("Other"))) %>%
  group_by(Family_lab) %>%
  summarise(abundance= n()) %>%
  mutate(database=paste("DonorB")) %>%
  mutate(Family_lab= factor(Family_lab, family_order)) %>%
ggplot(aes(database, abundance, fill=Family_lab)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = family_col) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_blank())
ggsave("../figs/MAGs_donorB.png",
       width = 3, height = 8, units = "cm")

##############################################################################
## 2. number of engrafted MAGs per patient
tbl %>%
  group_by(Fig_lab, DB, detection) %>%
  summarise(n=n()) %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) %>%
  filter(DB == "BDB" & detection == "Post") -> tbl0

# adding all patients (even those with no engraftment)
mapfile %>%
  filter(Treatment %in% c("FMT", "Placebo")) %>%
  select(Fig_lab, Treatment, Remission) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) %>%
  distinct() %>% 
  left_join(tbl0) %>%
  mutate(n= case_when(is.na(n) ~ paste(0),
                      TRUE ~ paste(n))) %>%
  mutate(detection= case_when(is.na(detection) ~ paste("Post"),
                      TRUE ~ paste(detection)))-> tbl1
tbl1$n <- as.numeric(tbl1$n)

tbl1 %>%
  arrange(-n) %>% pull(Fig_lab) -> pt_order

tbl1$Fig_lab <- factor(tbl1$Fig_lab, levels = pt_order)

ggplot(tbl1, aes(n, Fig_lab, fill=detection)) +
  geom_bar(stat = "identity") +
  facet_grid(Treatment~., space = "free", scales = "free") +
  scale_y_reordered() +
  scale_fill_manual(values = cols) +
  theme_classic() +
  theme(axis.title = element_blank(),
        legend.position = "none",
        strip.text = element_blank())
ggsave("../figs/MAGs_perfect_engrafted_counts.png",
       width = 4, height = 8, units = "cm")

##############################################################################
## 3. heatmap of engrafted MAGs per patients group by taxonomic families
tbl %>%
  filter(detection == "Post") %>%
  filter(DB== "BDB") %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) %>%
  left_join(gtdb_viz) %>%
  mutate(Family_lab= case_when(Family %in% tfam10 ~ paste(Family),
                               TRUE ~ paste("Other"))) %>%
  select(Fig_lab, bin_id, detection:Remission, Kingdom:Family_lab)-> tbl0

# adding all patients (even those with no engraftment)
mapfile %>%
  filter(Treatment %in% c("FMT", "Placebo")) %>%
  select(Fig_lab, Treatment, Remission) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) %>%
  distinct() %>% 
  left_join(tbl0) %>%
  mutate(Family_lab = case_when(is.na(Family_lab) ~ paste("Other"),
                                TRUE ~ paste(Family_lab)))-> tbl1

tbl1$Fig_lab <- factor(tbl1$Fig_lab, levels = pt_order)
tbl1$Family_lab <- factor(tbl1$Family_lab, levels = family_order)

ggplot(tbl1, aes(bin_id, Fig_lab, fill=detection)) +
  geom_tile() +
  facet_grid(Treatment~Family_lab, space = "free", scales= "free") +
  theme_classic() +
  scale_fill_manual(values = c("Post" = "#008837"), na.value = "white") +
  theme(axis.title = element_blank(), legend.position = "none",
        axis.text = element_blank(),
        strip.text.x = element_blank())
ggsave("../figs/MAGs_perfect_engrafted_heatmap.png",
       width = 12, height = 8, units = "cm")

```

## Comm engrafted MAGs

```{r}

C=90
table_P %>%
  ungroup() %>%
  mutate(sample= gsub("sub_", "", sample)) %>%
  mutate(Study_ID= gsub("_.*", "", sample)) %>%
  left_join(mapfile %>% select(Study_ID, Fig_lab, Timepoint)) %>%
  select(Fig_lab, bin_id, bin_coverage, DB, Timepoint) %>%
  spread(Timepoint, bin_coverage) %>%
  mutate(detection= case_when(WK0 >= C & WK6 >= C ~ paste("Both"),
                              WK0 >= C & WK6 < C ~ paste("Pre"),
                              WK0 < C & WK6 >= C ~ paste("Post"),
                              WK0 < C & WK6 < C ~ paste("none"),
                              TRUE ~ paste("Other"))) -> tbl

###############################################################################

common_Emagers <- function(tbl){
tbl %>%
  filter(detection == "Post") %>%
  select(Fig_lab, bin_id, WK6) %>%
  spread(Fig_lab, WK6, fill=0) -> engraft

# number of samples in input table
  tbl %>%
    pull(Fig_lab) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i),
                             E.mags = length(Emark.1[[i]]$bin_id))
  }
  do.call(rbind.data.frame, Emark.2)
}

###############################################################################

############################################
# common engraftment for Placebo
tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment) %>%
            distinct()) %>%
  filter(DB=="BDB") %>%
  filter(Treatment == "Placebo") -> tbl_pla
common_Emagers(tbl_pla) -> tbl_pla

############################################
# common engraftment for FMT
tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment) %>%
            distinct()) %>%
  filter(DB=="BDB") %>%
  filter(Treatment == "FMT") -> tbl_fmt
common_Emagers(tbl_fmt) -> tbl_fmt

############################################
# common engraftment for FMT-NoRes
tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  filter(DB=="BDB") %>%
  filter(Treatment == "FMT" & Remission == "NoRes") -> tbl_fmt_n
common_Emagers(tbl_fmt_n) -> tbl_fmt_n

############################################
# common engraftment for FMT-Res
tbl %>%
  left_join(mapfile %>% 
            select(Fig_lab, Treatment, Remission) %>%
            distinct()) %>%
  filter(DB=="BDB") %>%
  filter(Treatment == "FMT" & Remission == "Res") -> tbl_fmt_r
common_Emagers(tbl_fmt_r) -> tbl_fmt_r

######################################################
########### visualizing
######################################################

cols1 <- c("FMT(n=12)" = "#008837",
           "Placebo(n=12)" = "#bdbdbd",
           "Responder(n=6)" = "#0571b0",
           "Non-Responder(n=6)" = "#ca0020")

tbl_fmt %>% 
  mutate(Type=paste("FMT(n=12)"),
          Samples= "All") %>%
  bind_rows(tbl_pla %>% 
             mutate(Type=paste("Placebo(n=12)"),
                    Samples="All")) %>%
  bind_rows(tbl_fmt_r %>%
              mutate(Type="Responder(n=6)",
                     Samples="onlyFMT")) %>%
  bind_rows(tbl_fmt_n %>%
              mutate(Type="Non-Responder(n=6)",
                     Samples="onlyFMT")) -> tbl2
tbl2$Type <- factor(tbl2$Type, levels = c("FMT(n=12)", "Placebo(n=12)", 
                                        "Responder(n=6)", "Non-Responder(n=6)"))
ggplot(tbl2, aes(as.numeric(Npts), E.mags, color=Type)) +
  geom_line(aes(group=Type))+
  #facet_grid(~Samples, space = "free") +
  scale_y_sqrt() +
  scale_x_continuous(breaks = c(1, 2, 3, 4, 5, 6, 7,8,9,10)) +
  scale_color_manual(values = cols1) +
  theme_classic() +
  theme(legend.title = element_blank(),
        legend.position = c("0.8", "0.8")) +
  xlab("Increasing # of patients") +
  ylab("# of engrafted MAGs")
ggsave("../figs/MAG_CEG_curve.png", width = 8, height = 8, units = "cm")


#######################################################################
########### MAGs commonly engrafted in >= 3 pts within FMT and Placebo
#######################################################################

tbl %>%
  #only donor B MAGs (removinf UHGG)
  filter(!grepl("UHGG", bin_id)) %>%
  left_join(mapfile %>% select(Fig_lab, Treatment) %>%
              distinct()) %>%
  filter(detection == "Post" & Treatment == "FMT") %>%
  select(Fig_lab, bin_id, WK6) %>%
  spread(Fig_lab, WK6, fill=0) -> eng_fmt
engraft_3fmt <- eng_fmt[rowSums(eng_fmt[,2:ncol(eng_fmt)] > 0) >= 3, ]
engraft_3fmt %>%
  gather(Fig_lab, detection, -bin_id) %>%
  filter(detection >= 90) -> engraft_3fmt

tbl %>%
  #only donor B MAGs (removinf UHGG)
  filter(!grepl("UHGG", bin_id)) %>%
  left_join(mapfile %>% select(Fig_lab, Treatment) %>%
              distinct()) %>%
  filter(detection == "Post" & Treatment == "Placebo") %>%
  select(Fig_lab, bin_id, WK6) %>%
  spread(Fig_lab, WK6, fill=0) -> eng_placebo
engraft_3placebo <- eng_placebo[rowSums(eng_placebo[,2:ncol(eng_placebo)] > 0) >= 3, ]
engraft_3placebo %>%
  gather(Fig_lab, detection, -bin_id) %>%
  filter(detection >= 90) -> engraft_3placebo

engraft_3fmt %>%
  bind_rows(engraft_3placebo) %>%
  left_join(mapfile %>% select(Fig_lab, Treatment, Remission) %>%
              distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> engraft_3pts

# adding all patients (even those with no engraftment)
mapfile %>%
  filter(Treatment %in% c("FMT", "Placebo")) %>%
  select(Fig_lab, Treatment, Remission) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) %>%
  distinct() %>%
  left_join(engraft_3pts) %>%
  left_join(gtdb_viz) %>%
  # SINCE i ADDED ALL THE PATIENTS HERE (INCLUDING THOSE PLACEBO WITH NO 
  # ENGRAFTMETN) BECAUSE OF VISUALIZING. Here I'll manually change those
  # NA values so it looks alright visually.
  mutate(detection= case_when(is.na(detection) ~ paste("empty"),
                              TRUE ~ paste("Post"))) %>%
  # add the empty values to a bin just to look alright visually
  mutate(bin_id= case_when(detection == "empty" ~ paste("bin_219"),
                           TRUE ~ paste(bin_id))) %>%
  mutate(Genus= case_when(detection == "empty" ~ paste("UBA11774"),
                          TRUE ~ paste(Genus)))-> tbl3


ggplot(tbl3, aes(bin_id, Fig_lab, fill=detection)) +
  geom_tile() +
  facet_grid(Treatment~Genus, space = "free", scales = "free") +
  scale_fill_manual(values = c("Post" = "#008837", "empty" = "white")) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_text(size=5),
        strip.text.x = element_text(angle = 90, size = 5),
        legend.position = "none",
        axis.title = element_blank())
ggsave("../figs/MAG_CEG_heatmap.png", width = 12, height = 8, units = "cm")

```
