---
title: "paper figs"
author: "Sharok"
date: "2024-02-04"
output: html_document
---

make paper figures

```{r}

library(tidyverse)

```

Load main functions:
```{r}

source('./permutation_functions.R')

```

import permutation results (pvals for figs)
```{r}

perTest <- read.csv("../results/permuation_pvals.csv")
perTest %>%
  mutate(plot= case_when(str_detect(X, "FMTvPl") ~ paste("f1"),
                         str_detect(X, "ResvNoRes") ~ paste("f3"),
                         TRUE ~ paste("other"))) %>%
  mutate(data= case_when(str_detect(X, "16s") ~ paste("16s"),
                         str_detect(X, "species") ~ paste("species"),
                         str_detect(X, "strain") ~ paste("strain"),
                         str_detect(X, "mags") ~ paste("mags"),
                         str_detect(X, "genes") ~ paste("genes"),
                         )) %>%
  mutate(label= paste("Permutation Test \n p-value=", 
                      PWeightedEvents, sep = "")) -> perTest

```



function to make the main engraftment figure (f):

```{r}

f_paper <- function(f1, f2, f3, tx_pv, rs_pv, pvals) {
  
  # f1, f2, and f3 are the ggplot object 
  # generated by plot_fig13() and plot_fig2()
  # tx_pv and rs_pv named vectors of patient IDs/groups
  #pvals a datafame contains labels for f1 and f3 figures
  
  fig1_cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999")
 
   
 data.frame(element = tx_pv) %>%
 group_by(element) %>%
 tally() %>%
 mutate(annot= paste(element, "(n=", n, ")", sep = "")) %>%
 rename(Treatment="element") %>%
 select(Treatment, annot)-> i_tx
 #pull out ggplot data
 data1 <- ggplot_build(f1)$plot$data
 # re-label the color variable that include n
 labels <- setNames(i_tx$annot, i_tx$Treatment)
  # pvalue-label
  p_label = pvals %>% filter(plot == "f1") %>% pull(label)
  p1 = f1 +
  scale_color_manual(values = fig1_cols,
                     labels = labels) +
  annotate("text", 
           x = max(data1$Npats) * 0.85, y = max(data1$Count) * 0.4, 
           label = p_label, size = 2.5,
           colour = "black", parse = FALSE) +
  theme_classic() +
  theme(legend.position = c(0.75, 0.8),
        legend.title = element_blank(),
        axis.title = element_blank())
  #########################################
  fig2_cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999",
          "Both" = "#fc8d59")
  #pull out ggplot data
  data2 <- ggplot_build(f2)$plot$data
  # a label for second figure, showing n
  labels <- sprintf("FMT+Placebo\n(n=%d)", length(tx_pv))

  p2 = f2 +
  scale_fill_manual(values = fig2_cols) +
  scale_colour_manual(values = fig2_cols) +
  annotate("text", 
           x = max(data2$Total) * 0.8, y = max(data1$Count) * 0.8, 
           label = labels, size = 3,
           colour = "black", parse = FALSE) +
  theme_classic() +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.title = element_blank())
  #########################################
  fig3_cols <- c("Res" = "#2166ac",
          "NoRes" = "#b2182b")
  
 data.frame(element = rs_pv) %>%
 group_by(element) %>%
 tally() %>%
 mutate(annot= paste(element, "(n=", n, ")", sep = "")) %>%
 rename(Remission="element") %>%
 select(Remission, annot)-> i_rs
  #pull out ggplot data
 data3 <- ggplot_build(f3)$plot$data
 # re-label the color variable that include n
 labels <- setNames(i_rs$annot, i_rs$Remission)
  # pvalue-label
  p_label = pvals %>% filter(plot == "f3") %>% pull(label)
  p3 = f3 +
  scale_colour_manual(values = fig3_cols,
                      labels = labels) +
  annotate("text", 
           x = max(data3$Npats) * 0.85, y = max(data3$Count) * 0.4, 
           label = p_label, size = 2.5,
           colour = "black", parse = FALSE) +
  theme_classic() +
  theme(legend.position = c(0.75, 0.8),
        legend.title = element_blank(),
        axis.title = element_blank())
  
  cowplot::plot_grid(p1, p2, p3, ncol = 1)
  
}


```



## 16s data:

```{r}
# Import 16s mapfile and create named vectors of patient IDs/groups for permuting
mapfile <- read.csv("../data/UCFMT1_16Smap.txt")
tx_pv = pat_to_vect(mapfile, 'Fig_lab', 'Treatment')
rs_pv = pat_to_vect(filter(mapfile, Treatment == 'FMT'), 'Fig_lab', 
                          'Remission')
# Import the 16s ASV data
marker_lvl_16s <- read.csv("../data/16s_lvl_RA.csv")

# Import donor B 16s ASVs and create a vector
strains16s <- read.csv("../data/jcsz_16s_core_DonB.csv")
strains16s %>%
  pull() -> strains16s

# Set the cutoffs for 16s presence and absence
cutoff_pres_16S = 0.0001
cutoff_abs_16S = 0

### Get the engraftment matrix ###

# convert the 16s data to long format
long_mark_16s = mark_to_long(marker_lvl_16s, strains16s)
# Create a matrix of engraftment counts per patient
engr_16s = get_engraft(long_mark_16s, mapfile, cutoff_abs_16S, cutoff_pres_16S,
                       Treatment %in% c('FMT', 'Placebo'))

#### Fig 1 
cts_16s_plt = count_engraft(engr_16s, tx_pv, 'Treatment')
f1_16s = plot_fig_13(cts_16s_plt, 'Treatment')

#### Fig 2 
f2_16s = plot_fig_2(cts_16s_plt)

#### Fig 3
cts_16s_plt = count_engraft(engr_16s, rs_pv, 'Remission')
f3_16s = plot_fig_13(cts_16s_plt, 'Remission')


#### paper Fig
pvals_16s=perTest %>% filter(data == "16s")
f_paper(f1_16s, f2_16s, f3_16s, tx_pv, rs_pv, pvals_16s)
ggsave("../plots_paper/16s.pdf", height = 15, width = 7, units = "cm")

```

## species data

```{r}

#### Setup Metagenomics ####

# This mapfile is used for every non-16S feature type
mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

# These patient vectors are used for every non-16S feature type
tx_pv = pat_to_vect(mapfile, 'Fig_lab', 'Treatment')
rs_pv = pat_to_vect(filter(mapfile, Treatment == 'FMT'), 'Fig_lab', 
                          'Remission')

#### Species Setup ####

### Import the data

# Read in the species-level markers
marker_lvl_sp <- read.csv("../data/species_lvl_RA.csv")

# get a vector of donor B species
donorB_species = (marker_lvl_sp
                  %>% gather(sample, abundance, -Marker) 
                  #filtering donor B markers
                  %>% filter(str_detect(sample, "DonorB_D_")) 
                  # any marker with abundance > 0
                  # markers that were present in at least one donor B sample
                  %>% filter(!is.na(abundance)) 
                  %>% filter(as.numeric(abundance) > 0) 
                  %>% mutate(sample= gsub("DonorB_D_", "", sample)))

# Set the cutoff for "presence" with species. This is used for both patient
# samples and donor B
cutoff_pres_sp = 0.005

# Filter to only include donor B species that meet that cutoff
donorB_species = (donorB_species 
                  %>% filter(as.numeric(abundance) > cutoff_pres_sp))

# Turn it into a vector
donorB_species %>%
  select(Marker) %>% distinct() %>%
  pull() -> strains_sp

# Set the species absence cutoff
cutoff_abs_sp = 0

### Get the wide matrix features engrafted (1) or not (0) in patients
long_mark_sp = mark_to_long(marker_lvl_sp, strains_sp)
engr_sp = get_engraft(long_mark_sp, mapfile, cutoff_abs_sp, cutoff_pres_sp,
                       Treatment %in% c('FMT', 'Placebo'))
#### Species FMT vs Placebo ####

#### Fig 1
cts_sp = count_engraft(engr_sp, tx_pv, 'Treatment')
f1_sp = plot_fig_13(cts_sp, 'Treatment')

#### Fig 2 
f2_sp = plot_fig_2(cts_sp)

#### Species Res vs NoRes ####
cts_sp_rem = count_engraft(engr_sp, rs_pv, 'Remission')
f3_sp = plot_fig_13(cts_sp_rem, 'Remission')


#### paper Fig
pvals_sp=perTest %>% filter(data == "species")
f_paper(f1_sp, f2_sp, f3_sp, tx_pv, rs_pv, pvals_sp)
ggsave("../plots_paper/species.pdf", height = 15, width = 7, units = "cm")

```

## strains data

```{r}

#### Strains Setup ####

### Import the data

# Import the strain-level marker data
marker_lvl_st <- read.csv("../data/marker_lvl_RA.csv")

# Get the strains that are found in donor B
donorB_strains = (marker_lvl_st 
                  %>% gather(sample, abundance, -sp_marker, -st_marker, -Marker) 
                  %>% filter(str_detect(sample, "DonorB_D_"))
                  #filtering donor B markers
                  %>% filter(!is.na(abundance)) 
                  %>% filter(as.numeric(abundance) > 0) 
                  %>% mutate(sample= gsub("DonorB_D_", "", sample)))

# Set the presence cutoff used in both donor B and patient samples
cutoff_pres_st = 0.0001

# Filter donor B to only the strains that meet the cutoff
donorB_strains = (donorB_strains
                  %>% filter(as.numeric(abundance) > cutoff_pres_st))

# Turn it into a vector
donorB_strains %>%
  select(Marker) %>% distinct() %>%
  pull() -> strains_st

# Set the strain absence cutoff
cutoff_abs_st = 0

### Get the engraftment matrix

long_mark_st = mark_to_long(select(marker_lvl_st, -sp_marker, -st_marker), 
                            strains_st)
engr_st = get_engraft(long_mark_st, mapfile, cutoff_abs_st, cutoff_pres_st,
                       Treatment %in% c('FMT', 'Placebo'))


#### Strains FMT vs Placebo ####
###### f1
cts_st = count_engraft(engr_st, tx_pv, 'Treatment')
f1_st = plot_fig_13(cts_st, 'Treatment')

##### f2
f2_st = plot_fig_2(cts_st)

#### f3 
cts_st_rem = count_engraft(engr_st, rs_pv, 'Remission')
f3_st = plot_fig_13(cts_st_rem, 'Remission')

#### paper Fig
pvals_st=perTest %>% filter(data == "strain")
f_paper(f1_st, f2_st, f3_st, tx_pv, rs_pv, pvals_st)
ggsave("../plots_paper/strain.pdf", height = 15, width = 7, units = "cm")

```

## MAGs data

```{r}

#### MAGs Setup ####

### Import the data

# Import the MAG and bin data
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", header = F)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", header = F)
colnames(c_inBin) <- c("contig", "bin_id")
colnames(c_inMAG) <- c("contig", "bin_id")

# Get a vector of the mags
MAGs = (c_inMAG 
        %>% select(bin_id) 
        %>% distinct() 
        %>% pull())


## read in the MAG coverage
heads_cover <- c("sample.id",
                 "contig", "startpos", "endpos", 
                 "numreads", "covbases", "coverage", "meandepth",
                 "meanbaseq", "meanmapq",
                 "ReadSampling", "Quality")

path="../data/perfect_mapping/DBD"
patt="_perfect_.*.cover"
marker_lvl_mg = (data.frame(sample.id = paste(dir(path,
                                                  pattern = patt))) 
             # read each file from the directory (current path) and then unnest
                 %>% mutate(file_contents = map(sample.id, ~ 
                                                    read.csv(file.path(path, .),
                                                         sep = "\t", header = F,
                                                           comment.char = "#"))) 
                 %>% unnest() 
                 %>%  mutate(sample.id = gsub("_perfect_BDB.cover", 
                                              "", sample.id)))
colnames(marker_lvl_mg) <- heads_cover

# Join the coverage info with the bin info, filter to only keep mags, calculate
# normalized coverage
marker_lvl_mg = (marker_lvl_mg 
                 %>% left_join(c_inBin) 
                 %>% mutate(bin_q= case_when(bin_id %in% MAGs ~ paste("MAG"),
                                             TRUE ~ paste("Bin"))) 
                 %>% filter(bin_q == "MAG") 
                 %>% select(-bin_q) 
                 %>% dplyr::rename(sample=sample.id) 
                 %>% group_by(sample, bin_id) 
                 %>% summarise(bin_len= sum(endpos),
                               bin_covbases=sum(covbases),
                               bin_numreads=sum(numreads),
                               bin_coverage=bin_covbases/bin_len*100))

# Keep only the patient samples, reshape the numbers, and name the Marker column
marker_lvl_mg = (marker_lvl_mg 
                 %>% filter(str_detect(sample, "PMCL")) 
                 %>% select(sample, bin_id, bin_coverage) 
                 %>% spread(sample, bin_coverage) 
                 %>% rename(Marker= bin_id))

# Set the presence and absence cutoffs for patients. DonorB doesn't need these
# because all the MAGs come from donor B
cutoff_pres_mg = 90
cutoff_abs_mg = 25


### Get the engraftment matrix

long_mark_mg = mark_to_long(marker_lvl_mg)
engr_mg = get_engraft(long_mark_mg, mapfile, cutoff_abs_mg, cutoff_pres_mg,
                       Treatment %in% c('FMT', 'Placebo'))


##### f1
cts_mg = count_engraft(engr_mg, tx_pv, 'Treatment')
f1_mg = plot_fig_13(cts_mg, 'Treatment')

##### f2
f2_mg = plot_fig_2(cts_mg)

#### f3
cts_mg_rem = count_engraft(engr_mg, rs_pv, 'Remission')
f3_mg = plot_fig_13(cts_mg_rem,'Remission')

#### paper Fig
pvals_mg=perTest %>% filter(data == "mags")
f_paper(f1_mg, f2_mg, f3_mg, tx_pv, rs_pv, pvals_mg)
ggsave("../plots_paper/MAGs.pdf", height = 15, width = 7, units = "cm")


```

## Gene data

```{r}

#### Genes Setup ####

### Import the data

marker_lvl_ge <- read.csv("../data/genes_lvl.csv")

# Set the presence and absence cutoffs for patients. These aren't used for donor
# B because all the genes came from the donor B assemblies.
cutoff_pres_ge = 90
cutoff_abs_ge = 25

### Get the engraftment matrix

long_mark_ge = mark_to_long(marker_lvl_ge)
engr_ge = get_engraft(long_mark_ge, mapfile, cutoff_abs_ge, cutoff_pres_ge,
                       Treatment %in% c('FMT', 'Placebo'))


#### f1
cts_ge = count_engraft(engr_ge, tx_pv, 'Treatment')
f1_ge = plot_fig_13(cts_ge, 'Treatment')

##### f2
f2_ge = plot_fig_2(cts_ge)

#investigate second figure later
#ggplot_build(f2_ge)$plot$data -> test

#### f3
cts_ge_rem = count_engraft(engr_ge, rs_pv, 'Remission')
f3_ge = plot_fig_13(cts_ge_rem, 'Remission')

#### paper Fig
pvals_ge=perTest %>% filter(data == "genes")
f_paper(f1_ge, f2_mg, f3_ge, tx_pv, rs_pv, pvals_ge)
ggsave("../plots_paper/genes.pdf", height = 15, width = 7, units = "cm")

```

