---
title: "paper figs"
author: "Sharok"
date: "2024-02-04"
output: html_document
---

make paper figures

```{r}

library(tidyverse)
library(tidytext)

```

Load main functions:
```{r}

source('./permutation_functions.R')
source('./paperfig_functions.R')

```

import permutation results (pvals for figs)
```{r}

perTest <- read.csv("../results/permuation_pvals.csv")
perTest %>%
  mutate(plot= case_when(str_detect(X, "FMTvPl") ~ paste("f1"),
                         str_detect(X, "ResvNoRes") ~ paste("f3"),
                         TRUE ~ paste("other"))) %>%
  mutate(data= case_when(str_detect(X, "16s") ~ paste("16s"),
                         str_detect(X, "species") ~ paste("species"),
                         str_detect(X, "strain") ~ paste("strain"),
                         str_detect(X, "mags") ~ paste("mags"),
                         str_detect(X, "genes") ~ paste("genes"),
                         )) %>%
  mutate(label= paste("Permutation Test \n p-value=", 
                      PWeightedEvents, sep = "")) -> perTest

```



function to make the main engraftment figure (f) moved to paperfig_functions.R


function to make engraftment profiles; Profiling donor's markers in patients.
Moved to paperfig_functions.R

## benchmarking function for 16s (relative abundance%)
notes to Jake: this function is exactly same as species/strain but just removed
the colors so that we don't have to have the same names for samples etc.
we can decide to maybe keep one function with no color for any of them

Moved to paperfig_functions.R

## 16s data:

```{r}
# Import 16s mapfile and create named vectors of patient IDs/groups for permuting
mapfile <- read.csv("../data/UCFMT1_16Smap.txt")
tx_pv = pat_to_vect(mapfile, 'Fig_lab', 'Treatment')
rs_pv = pat_to_vect(filter(mapfile, Treatment == 'FMT'), 'Fig_lab', 
                          'Remission')
# Import the 16s ASV data
marker_lvl_16s <- read.csv("../data/16s_lvl_RA.csv")

# Import donor B 16s ASVs and create a vector
strains16s <- read.csv("../data/jcsz_16s_core_DonB.csv")
strains16s %>%
  pull() -> strains16s


donorB_16s_samples <- read.csv("../data/16s_lvl_RA_donorB_samples.csv")
donorB_16s_samples %>%
  gather(sample, abundance, -Marker) %>%
 # any marker with abundance > 0
# markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) -> donorB_16s_samples

# Pick 4 random samples
set.seed(5)
bs_4 = (donorB_16s_samples
        %>% select(sample)
        %>% unique()
        %>% pull()
        %>% sample(size = 4, replace = FALSE))

# benchmarking figures
breaks_sp <- c(0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1)
benchmark(donorB_16s_samples, breaks_sp, f_type = '16s', bs_4)
ggsave("../plots_paper/16s_cutoff.svg", height = 12, width = 10, units = "cm")



# Set the cutoffs for 16s presence and absence
cutoff_pres_16S = 0.00013
cutoff_abs_16S = 0

### Get the engraftment matrix ###

# convert the 16s data to long format
long_mark_16s = mark_to_long(marker_lvl_16s, strains16s)
# Create a matrix of engraftment counts per patient
engr_16s = get_engraft(long_mark_16s, mapfile, cutoff_abs_16S, cutoff_pres_16S,
                       Treatment %in% c('FMT', 'Placebo'))

# profile the feature types:
profile_16s = get_profile_pat(long_mark_16s, mapfile, cutoff_abs_16S, cutoff_pres_16S,
                       Treatment %in% c('FMT', 'Placebo'))
profile_16s_don = get_profile_don(long_mark_16s, mapfile, cutoff_abs_16S,
                                  cutoff_pres_16S, Treatment %in% c('FMT', 'Placebo'))

p_paper(profile_16s, profile_16s_don, mapfile)
ggsave("../plots_paper/16s_profile.svg", height = 15, width = 22.5, units = "cm")

#### Fig 1 
cts_16s_plt = count_engraft(engr_16s, tx_pv, 'Treatment')
f1_16s = plot_fig_13(cts_16s_plt, 'Treatment')

#### Fig 2 
f2_16s = plot_fig_2(cts_16s_plt)

#### Fig 3
cts_16s_plt = count_engraft(engr_16s, rs_pv, 'Remission')
f3_16s = plot_fig_13(cts_16s_plt, 'Remission')


#### paper Fig
pvals_16s=perTest %>% filter(data == "16s")
f_paper(f1_16s, f2_16s, f3_16s, tx_pv, rs_pv, pvals_16s)
ggsave("../plots_paper/16s.svg", height = 15, width = 7, units = "cm")

```



## benchmarking function for species and strains (relative abundance%) moved to
paperfig_functions.R

## species data

```{r}

#### Setup Metagenomics ####

# This mapfile is used for every non-16S feature type
mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

# These patient vectors are used for every non-16S feature type
tx_pv = pat_to_vect(mapfile, 'Fig_lab', 'Treatment')
rs_pv = pat_to_vect(filter(mapfile, Treatment == 'FMT'), 'Fig_lab', 
                          'Remission')

#### Species Setup ####

### Import the data

# Read in the species-level markers
marker_lvl_sp <- read.csv("../data/species_lvl_RA.csv")

# get a vector of donor B species
donorB_species = (marker_lvl_sp
                  %>% gather(sample, abundance, -Marker) 
                  #filtering donor B markers
                  %>% filter(str_detect(sample, "DonorB_D_")) 
                  # any marker with abundance > 0
                  # markers that were present in at least one donor B sample
                  %>% filter(!is.na(abundance)) 
                  %>% filter(as.numeric(abundance) > 0) 
                  %>% mutate(sample= gsub("DonorB_D_", "", sample),
                             abundance = abundance / 100))

# benchmarking figures
breaks_sp <- c(0.00001, 0.00005, 0.0001, 0.001, 0.01, 0.02, 0.03)
benchmark(donorB_species, breaks_sp, f_type = 'sp')
ggsave("../plots_paper/species_cutoff.svg", height = 12, width = 10, units = "cm")

# Set the cutoff for "presence" with species. This is used for both patient
# samples and donor B
cutoff_pres_sp = 0.000073

# Filter to only include donor B species that meet that cutoff
donorB_species = (donorB_species 
                  %>% filter(as.numeric(abundance) > cutoff_pres_sp))

# Turn it into a vector
donorB_species %>%
  select(Marker) %>% distinct() %>%
  pull() -> strains_sp

# Set the species absence cutoff
cutoff_abs_sp = 0

### Get the wide matrix features engrafted (1) or not (0) in patients
long_mark_sp = mark_to_long(marker_lvl_sp, strains_sp)
engr_sp = get_engraft(long_mark_sp, mapfile, cutoff_abs_sp, cutoff_pres_sp,
                       Treatment %in% c('FMT', 'Placebo'))

# profile the feature types:
profile_sp = get_profile_pat(long_mark_sp, mapfile, cutoff_abs_sp, cutoff_pres_sp,
                       Treatment %in% c('FMT', 'Placebo'))
profile_sp_don = get_profile_don(long_mark_sp, mapfile, cutoff_abs_sp,
                                  cutoff_pres_sp, Treatment %in% c('FMT', 'Placebo'))

p_paper(profile_sp, profile_sp_don, mapfile)
ggsave("../plots_paper/species_profile.svg", height = 15, width = 22.5, 
       units = "cm")

#### Species FMT vs Placebo ####

#### Fig 1
cts_sp = count_engraft(engr_sp, tx_pv, 'Treatment')
f1_sp = plot_fig_13(cts_sp, 'Treatment')

#### Fig 2 
f2_sp = plot_fig_2(cts_sp)

#### Species Res vs NoRes ####
cts_sp_rem = count_engraft(engr_sp, rs_pv, 'Remission')
f3_sp = plot_fig_13(cts_sp_rem, 'Remission')

#### paper Fig
pvals_sp=perTest %>% filter(data == "species")
f_paper(f1_sp, f2_sp, f3_sp, tx_pv, rs_pv, pvals_sp)
ggsave("../plots_paper/species.svg", height = 15, width = 7, units = "cm")

```

## strains data

```{r}

#### Strains Setup ####

### Import the data

# Import the strain-level marker data
marker_lvl_st <- read.csv("../data/marker_lvl_RA.csv")

# Get the strains that are found in donor B
donorB_strains = (marker_lvl_st 
                  %>% gather(sample, abundance, -sp_marker, -st_marker, -Marker) 
                  %>% filter(str_detect(sample, "DonorB_D_"))
                  #filtering donor B markers
                  %>% filter(!is.na(abundance)) 
                  %>% filter(as.numeric(abundance) > 0) 
                  %>% mutate(sample= gsub("DonorB_D_", "", sample),
                             abundance = abundance / 100))

# benchmarking figures
breaks_st <- c(0.00000001, 0.0000001, 0.000001, 0.00001, 0.0001, 0.001, 0.01)
benchmark(donorB_strains, breaks_st, f_type = 'sp')
ggsave("../plots_paper/strains_cutoff.svg", height = 12, width = 10, units = "cm")


# Set the presence cutoff used in both donor B and patient samples
cutoff_pres_st = 0.0000013

# Filter donor B to only the strains that meet the cutoff
donorB_strains = (donorB_strains
                  %>% filter(as.numeric(abundance) > cutoff_pres_st))

# Turn it into a vector
donorB_strains %>%
  select(Marker) %>% distinct() %>%
  pull() -> strains_st

# Set the strain absence cutoff
cutoff_abs_st = 0

### Get the engraftment matrix

long_mark_st = mark_to_long(select(marker_lvl_st, -sp_marker, -st_marker), 
                            strains_st)
engr_st = get_engraft(long_mark_st, mapfile, cutoff_abs_st, cutoff_pres_st,
                       Treatment %in% c('FMT', 'Placebo'))

# profile the feature types:
profile_st = get_profile_pat(long_mark_st, mapfile, cutoff_abs_st, cutoff_pres_st,
                       Treatment %in% c('FMT', 'Placebo'))
profile_st_don = get_profile_don(long_mark_st, mapfile, cutoff_abs_st,
                                  cutoff_pres_st, Treatment %in% c('FMT', 'Placebo'))

p_paper(profile_st, profile_st_don, mapfile)
ggsave("../plots_paper/strains_profile.svg", height = 15, width = 22.5, units = "cm")

#### Strains FMT vs Placebo ####
###### f1
cts_st = count_engraft(engr_st, tx_pv, 'Treatment')
f1_st = plot_fig_13(cts_st, 'Treatment')

##### f2
f2_st = plot_fig_2(cts_st)

#### f3 
cts_st_rem = count_engraft(engr_st, rs_pv, 'Remission')
f3_st = plot_fig_13(cts_st_rem, 'Remission')

#### paper Fig
pvals_st=perTest %>% filter(data == "strain")
f_paper(f1_st, f2_st, f3_st, tx_pv, rs_pv, pvals_st)
ggsave("../plots_paper/strain.svg", height = 15, width = 7, units = "cm")

```


## MAGs data

```{r}

#### MAGs Setup ####

### Import the data

# Import the MAG and bin data
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", header = F)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", header = F)
colnames(c_inBin) <- c("contig", "bin_id")
colnames(c_inMAG) <- c("contig", "bin_id")

# Get a vector of the mags
MAGs = (c_inMAG 
        %>% select(bin_id) 
        %>% distinct() 
        %>% pull())

# gtdb taxonomy output for all the bins
gtdb.bac <- read.csv("../data/Assembly/gtdbtk.bac120.summary.tsv", sep = "\t")
gtdb.arc <- read.csv("../data/Assembly/gtdbtk.ar122.summary.tsv", sep = "\t")

gtdb.bac %>% mutate(bin_id=user_genome) %>%
  select(bin_id, classification) %>%
  bind_rows(gtdb.arc %>% mutate(bin_id=user_genome) %>%
              select(bin_id, classification)) %>%
  mutate(bin_id=gsub("bin.", "bin_", bin_id)) -> gtdb



## read in the MAG coverage
heads_cover <- c("sample.id",
                 "contig", "startpos", "endpos", 
                 "numreads", "covbases", "coverage", "meandepth",
                 "meanbaseq", "meanmapq",
                 "ReadSampling", "Quality")

path="../data/perfect_mapping/DBD"
patt="_perfect_.*.cover"
marker_lvl_mg = (data.frame(sample.id = paste(dir(path,
                                                  pattern = patt))) 
             # read each file from the directory (current path) and then unnest
                 %>% mutate(file_contents = map(sample.id, ~ 
                                                    read.csv(file.path(path, .),
                                                         sep = "\t", header = F,
                                                           comment.char = "#"))) 
                 %>% unnest() 
                 %>%  mutate(sample.id = gsub("_perfect_BDB.cover", 
                                              "", sample.id)))
colnames(marker_lvl_mg) <- heads_cover

# Join the coverage info with the bin info, filter to only keep mags, calculate
# normalized coverage
marker_lvl_mg = (marker_lvl_mg 
                 %>% left_join(c_inBin) 
                 %>% mutate(bin_q= case_when(bin_id %in% MAGs ~ paste("MAG"),
                                             TRUE ~ paste("Bin"))) 
                 %>% filter(bin_q == "MAG") 
                 %>% select(-bin_q) 
                 %>% dplyr::rename(sample=sample.id) 
                 %>% group_by(sample, bin_id) 
                 %>% summarise(bin_len= sum(endpos),
                               bin_covbases=sum(covbases),
                               bin_numreads=sum(numreads),
                               bin_coverage=bin_covbases/bin_len))

marker_lvl_mg %>%
  filter(str_detect(sample, "DonorB_D_")) %>%
  select(sample, bin_id, bin_coverage) %>%
  rename(Genome=bin_id, coverage=bin_coverage) %>%
  mutate(sample= gsub("DonorB_D_", "", sample))-> donorB_mag


# benchmarking figures
breaks_mg <- c(0, 0.25, 0.50, 0.75, 1)
Genome_benchmark(donorB_mag, breaks_mg, 'MAG')
ggsave("../plots_paper/MAGs_cutoff.svg", height = 12, width = 10, units = "cm")

# Keep only the patient samples, reshape the numbers, and name the Marker column
marker_lvl_mg = (marker_lvl_mg 
                 %>% filter(str_detect(sample, "PMCL")) 
                 %>% select(sample, bin_id, bin_coverage) 
                 %>% spread(sample, bin_coverage) 
                 %>% rename(Marker= bin_id))

# Set the presence and absence cutoffs for patients. DonorB doesn't need these
# because all the MAGs come from donor B
cutoff_pres_mg = 0.75
cutoff_abs_mg = 0.20


### Get the engraftment matrix

long_mark_mg = mark_to_long(marker_lvl_mg)
engr_mg = get_engraft(long_mark_mg, mapfile, cutoff_abs_mg, cutoff_pres_mg,
                       Treatment %in% c('FMT', 'Placebo'))

# profile the feature types:
profile_mg = get_profile_pat(long_mark_mg, mapfile, cutoff_abs_mg, cutoff_pres_mg,
                       Treatment %in% c('FMT', 'Placebo'))
profile_mg_don = get_profile_don(long_mark_mg, mapfile, cutoff_abs_mg, cutoff_pres_mg,
                                 Treatment %in% c('FMT', 'Placebo'))
p_paper(profile_mg, profile_mg_don, mapfile)
ggsave("../plots_paper/MAGs_profile.svg", height = 15, width = 22.5, units = "cm")


##### f1
cts_mg = count_engraft(engr_mg, tx_pv, 'Treatment')
f1_mg = plot_fig_13(cts_mg, 'Treatment')

##### f2
f2_mg = plot_fig_2(cts_mg)

#### f3
cts_mg_rem = count_engraft(engr_mg, rs_pv, 'Remission')
f3_mg = plot_fig_13(cts_mg_rem,'Remission')

#### paper Fig
pvals_mg=perTest %>% filter(data == "mags")
f_paper(f1_mg, f2_mg, f3_mg, tx_pv, rs_pv, pvals_mg)
ggsave("../plots_paper/MAGs.svg", height = 15, width = 7, units = "cm")


```

Is there a signature of MAGs specific to responder?

```{r}

gtdb %>%
  separate(classification, c("Kingdom", "Phylum", "Class",
                                   "Order", "Family", "Genus",
                                   "Species"), ";.__") -> gtdb_viz
#select top 10 families
gtdb_viz %>%
  rename(Marker = bin_id) %>%
  filter(Marker %in% MAGs) %>%
  select(Marker, Family) %>%
  group_by(Family) %>%
  summarise(n= n()) %>%
  arrange(-n) %>% pull(Family) -> tfam
tfam10 <- tfam[1:10]

##############################################################################
## 1. summary of taxonomy of donor B MAGs
##############################################################################
family_col= c(
"Lachnospiraceae" = "#6a3d9a",
"Ruminococcaceae" = "#33a02c",
"Bacteroidaceae" = "#1f78b4",
"Oscillospiraceae" = "#ff7f00",
"CAG-508" = "#b2df8a",
"Acutalibacteraceae" = "#fdbf6f",
"Streptococcaceae" = "#ffff99",
"CAG-822" =  "#a6cee3",
"Lactobacillaceae" = "#fb9a99",
"Butyricicoccaceae" = "#f781bf",
"Other" = "#808080"
)

family_order <- c("Lachnospiraceae", "Oscillospiraceae",
                  "Ruminococcaceae", "CAG-508", "Bacteroidaceae", 
                  "Acutalibacteraceae", "Lactobacillaceae",
                  "Streptococcaceae", "Butyricicoccaceae",
                  "CAG-822", "Other")

gtdb_viz %>%
mutate(Family_lab= case_when(Family %in% tfam10 ~ paste(Family),
                               TRUE ~ paste("Other"))) -> gtdb_viz


# select the marker that were engrafted in >= 3 patients
data.frame(cts_mg) %>%
  mutate(Total = rowSums(.),
         Uniqueness = case_when(tx_tot > 0 & bl_tot > 0 ~ 'Both',
                                tx_tot == 0 & bl_tot > 0 ~ 'Placebo',
                                tx_tot > 0 & bl_tot == 0 ~ 'FMT',
                                tx_tot == 0 & bl_tot == 0 ~ 'None')) %>%
  filter(Total >= 1) %>%
  rownames_to_column("Marker") %>%
  pull(Marker)-> of_int_marker
# look for these marker of interest across all patients
data.frame(engr_mg) %>%
  rownames_to_column("Marker") %>%
  filter(Marker %in% of_int_marker) %>%
  column_to_rownames("Marker") -> engr_mg_2

# f1 data with only marker of int
cts_mg_2 = count_engraft(engr_mg_2, tx_pv, 'Treatment')
data.frame(cts_mg_2) %>%
  rename(FMT=tx_tot,
         Placebo=bl_tot) %>%
  rownames_to_column("Marker") -> cts_mg_2
# f3 data with only marker of int
cts_mg_rem_2 = count_engraft(engr_mg_2, rs_pv, 'Remission')
data.frame(cts_mg_rem_2) %>%
  rename(Res=tx_tot,
         NoRes=bl_tot) %>%
  rownames_to_column("Marker") -> cts_mg_rem_2

cts_mg_2 %>%
  left_join(cts_mg_rem_2) %>%
  mutate(tUniq = case_when(FMT > 0 & Placebo > 0 ~ 'Both',
                                FMT == 0 & Placebo > 0 ~ 'Placebo',
                                FMT > 0 & Placebo == 0 ~ 'FMT',
                                FMT == 0 & Placebo == 0 ~ 'None')) %>%
  mutate(rUniq = case_when(tUniq == "FMT" & Res > 0 & NoRes > 0 ~ 'F-Both',
                           tUniq == "FMT" & Res == 0 & NoRes > 0 ~ 'F-NoRes',
                           tUniq == "FMT" & Res > 0 & NoRes == 0 ~ 'F-Res',
                           tUniq == "FMT" & Res == 0 & NoRes == 0 ~ 'F-None',
                           tUniq == "Placebo" & Res > 0 & NoRes > 0 ~ 'P-Both',
                           tUniq == "Placebo" & Res == 0 & NoRes > 0 ~ 'P-NoRes',
                           tUniq == "Placebo" & Res > 0 & NoRes == 0 ~ 'P-Res',
                           tUniq == "Placebo" & Res == 0 & NoRes == 0 ~ 'P-None',
                           tUniq == "Both" ~ 'Both',
                           TRUE ~ 'None')) %>%
  select(Marker, tUniq, rUniq) -> Marker_Uniq


data.frame(engr_mg_2) %>%
 rownames_to_column("Marker") %>%
 gather(Fig_lab, engraft, -Marker) %>%
 filter(engraft > 0) %>%
 left_join(Marker_Uniq) %>%
 left_join(mapfile %>% select(Fig_lab, Treatment, Remission)) %>%
 mutate(Fig_lab = case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ Fig_lab))-> table

# adding MAGs taxonomy:
table %>%
  left_join(gtdb_viz %>% rename(Marker = bin_id)) %>%
  mutate(FamLab= paste(Family, Marker, sep = "-")) -> table


table$rUniq <- factor(table$rUniq, levels = c("Both", "P-None", "F-Both",
                                              "F-NoRes", "F-Res"))

fig2_cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999",
          "Both" = "#fc8d59")

ggplot(table, aes(Marker, Fig_lab, fill=tUniq)) +
  geom_tile() +
  facet_grid(Treatment~rUniq, space = "free", scales = "free") +
  scale_fill_manual(values = fig2_cols) +
  #scale_x_reordered() +
  theme_classic() +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title = element_blank(),
        legend.position = "none")
ggsave("../plots_paper/mags_ce.svg",
       width = 14, height = 6, units = "cm")


ggplot(table, aes(engraft, Fig_lab, fill=Family_lab)) +
  geom_bar(stat = "identity") +
  facet_grid(Treatment~rUniq, space = "free", scales = "free") +
  scale_fill_manual(values = family_col) +
  #scale_x_reordered() +
  theme_classic() +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        legend.position = "none")
ggsave("../plots_paper/mags_ce_taxa.svg",
       width = 14, height = 6, units = "cm")


table %>%
  select(Marker, rUniq) %>%
  distinct() %>%
  group_by(rUniq) %>%
  tally()

```


## UHGG MAGs

```{r}

marker_lvl_uhgg <- read.csv("../data/perfect_mapping/UHGG/UHGG_1x_cover_perfect_subset_reads.csv")

marker_lvl_uhgg %>%
  mutate(bin_coverage=bin_covbases/bin_len*100) %>%
  rename(sample=sample.id) %>%
  mutate(sample= gsub("_perfect_UHGG.cover", "", sample)) %>%
  mutate(sample = gsub("sub_", "", sample))-> marker_lvl_uhgg


marker_lvl_uhgg %>%
  filter(str_detect(sample, "DonorB_D_")) %>%
  select(sample, Genome, bin_coverage) %>%
  rename(coverage=bin_coverage) %>%
  mutate(sample= gsub("DonorB_D_", "", sample))-> donorB_uhgg

# benchmarking figures
breaks_mg <- c(0, 25, 50, 75, 100)
Genome_benchmark(donorB_uhgg, breaks_mg)
ggsave("../plots_paper/uhgg_cutoff.svg", height = 12, width = 10, units = "cm")

# Set the presence and absence cutoffs for patients. DonorB doesn't need these
# because all the MAGs come from donor B
cutoff_pres_mg = 75
cutoff_abs_mg = 25

# Keep only the donor samples samples,
marker_lvl_uhgg %>%
  filter(str_detect(sample, "DonorB_D_")) %>%
  filter(bin_coverage >= cutoff_pres_mg) %>%
  select(Genome) %>% distinct() %>%
  pull(Genome) -> donorB_inUHGG

# Keep only the patient samples, reshape the numbers, and name the Marker column
marker_lvl_uhgg = (marker_lvl_uhgg
                 %>% filter(str_detect(sample, "PMCL")) 
                 %>% select(sample, Genome, bin_coverage)
                 %>% filter(Genome %in% donorB_inUHGG)
                 %>% spread(sample, bin_coverage) 
                 %>% rename(Marker= Genome))

```



## Gene data

```{r}

#### Genes Setup ####

### Import the data

marker_lvl_ge <- read.csv("../data/genes_lvl.csv")

# Set the presence and absence cutoffs for patients. These aren't used for donor
# B because all the genes came from the donor B assemblies.
cutoff_pres_ge = 0.90
cutoff_abs_ge = 0.25

### Get the engraftment matrix

long_mark_ge = mark_to_long(marker_lvl_ge)
long_mark_ge = (long_mark_ge
                %>% mutate(abundance = abundance/100))
engr_ge = get_engraft(long_mark_ge, mapfile, cutoff_abs_ge, cutoff_pres_ge,
                       Treatment %in% c('FMT', 'Placebo'))

# profile the feature types:
profile_ge = get_profile_pat(long_mark_ge, mapfile, cutoff_abs_ge, cutoff_pres_ge,
                       Treatment %in% c('FMT', 'Placebo'))
profile_ge_don = get_profile_don(long_mark_ge, mapfile, cutoff_abs_ge, cutoff_pres_ge,
                                 Treatment %in% c('FMT','Placebo'))
p_paper(profile_ge, profile_ge_don, mapfile)
ggsave("../plots_paper/genes_profile.svg", height = 15, width = 22.5, units = "cm")

#### f1
cts_ge = count_engraft(engr_ge, tx_pv, 'Treatment')
f1_ge = plot_fig_13(cts_ge, 'Treatment')

##### f2
f2_ge = plot_fig_2(cts_ge)

#investigate second figure later
#ggplot_build(f2_ge)$plot$data -> test

#### f3
cts_ge_rem = count_engraft(engr_ge, rs_pv, 'Remission')
f3_ge = plot_fig_13(cts_ge_rem, 'Remission')

#### paper Fig
pvals_ge=perTest %>% filter(data == "genes")
f_paper(f1_ge, f2_ge, f3_ge, tx_pv, rs_pv, pvals_ge)
ggsave("../plots_paper/genes.svg", height = 15, width = 7, units = "cm")


```
A signature of genes specific to FMT responders:

```{r}

# select the marker that were engrafted in >= 3 patients
data.frame(cts_ge) %>%
  mutate(Total = rowSums(.),
         Uniqueness = case_when(tx_tot > 0 & bl_tot > 0 ~ 'Both',
                                tx_tot == 0 & bl_tot > 0 ~ 'Placebo',
                                tx_tot > 0 & bl_tot == 0 ~ 'FMT',
                                tx_tot == 0 & bl_tot == 0 ~ 'None')) %>%
  filter(Total >= 3) %>%
  rownames_to_column("Marker") %>%
  pull(Marker)-> of_int_marker
# look for these marker of interest across all patients
data.frame(engr_ge) %>%
  rownames_to_column("Marker") %>%
  filter(Marker %in% of_int_marker) %>%
  column_to_rownames("Marker") -> engr_ge_3

# f1 data with only marker of int
cts_ge_3 = count_engraft(engr_ge_3, tx_pv, 'Treatment')
data.frame(cts_ge_3) %>%
  rename(FMT=tx_tot,
         Placebo=bl_tot) %>%
  rownames_to_column("Marker") -> cts_ge_3
# f3 data with only marker of int
cts_ge_rem_3 = count_engraft(engr_ge_3, rs_pv, 'Remission')
data.frame(cts_ge_rem_3 ) %>%
  rename(Res=tx_tot,
         NoRes=bl_tot) %>%
  rownames_to_column("Marker") -> cts_ge_rem_3

cts_ge_3 %>%
  left_join(cts_ge_rem_3) %>%
  mutate(tUniq = case_when(FMT > 0 & Placebo > 0 ~ 'Both',
                                FMT == 0 & Placebo > 0 ~ 'Placebo',
                                FMT > 0 & Placebo == 0 ~ 'FMT',
                                FMT == 0 & Placebo == 0 ~ 'None')) %>%
  mutate(rUniq = case_when(tUniq == "FMT" & Res > 0 & NoRes > 0 ~ 'F-Both',
                           tUniq == "FMT" & Res == 0 & NoRes > 0 ~ 'F-NoRes',
                           tUniq == "FMT" & Res > 0 & NoRes == 0 ~ 'F-Res',
                           tUniq == "FMT" & Res == 0 & NoRes == 0 ~ 'F-None',
                           tUniq == "Placebo" & Res > 0 & NoRes > 0 ~ 'P-Both',
                           tUniq == "Placebo" & Res == 0 & NoRes > 0 ~ 'P-NoRes',
                           tUniq == "Placebo" & Res > 0 & NoRes == 0 ~ 'P-Res',
                           tUniq == "Placebo" & Res == 0 & NoRes == 0 ~ 'P-None',
                           tUniq == "Both" ~ 'Both',
                           TRUE ~ 'None')) %>%
  select(Marker, tUniq, rUniq) -> Marker_Uniq


data.frame(engr_ge_3) %>%
 rownames_to_column("Marker") %>%
 gather(Fig_lab, engraft, -Marker) %>%
 filter(engraft > 0) %>%
 left_join(Marker_Uniq) %>%
 left_join(mapfile %>% select(Fig_lab, Treatment, Remission)) %>%
 mutate(Fig_lab = case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ Fig_lab))-> table


table$rUniq <- factor(table$rUniq, levels = c("Both", "P-None", "F-Both",
                                              "F-NoRes", "F-Res"))

fig2_cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999",
          "Both" = "#fc8d59")

ggplot(table, aes(Marker, Fig_lab, fill=tUniq)) +
  geom_tile() +
  facet_grid(Treatment~rUniq, space = "free", scales = "free") +
  scale_fill_manual(values = fig2_cols) +
  #scale_x_reordered() +
  theme_classic() +
  theme(strip.text.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title = element_blank(),
        legend.position = "none")
ggsave("../plots_paper/genes_ce.svg",
       width = 14, height = 8, units = "cm")

# number of genes in each rUniq category (strips in x-axis)
table %>%
  select(Marker, rUniq) %>%
  distinct() %>%
  group_by(rUniq) %>% tally()

# save the list of genes specific to F-Res: 1488 genes
table %>%
  filter(rUniq == "F-Res") %>%
  distinct() %>%
  write.table("../data/genes_FMT-Res/gene_table.txt",
            row.names = F, quote = F, sep = "\t")

table %>%
  filter(rUniq == "F-Res") %>%
  distinct() %>%
  select(Marker) %>%
  write.table("../data/genes_FMT-Res/gene_ids.txt",
            row.names = F, col.names = F, quote = F, sep = "\t")

```

gene signature in bins

```{r}

# map information for contigs and bins
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", header = F)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", header = F)
colnames(c_inBin) <- c("contig", "bin_id")
colnames(c_inMAG) <- c("contig", "bin_id")

c_inMAG %>% select(bin_id) %>% distinct() %>% pull() -> MAGs


# gtdb taxonomy output for all the bins
gtdb.bac <- read.csv("../data/Assembly/gtdbtk.bac120.summary.tsv", sep = "\t")
gtdb.arc <- read.csv("../data/Assembly/gtdbtk.ar122.summary.tsv", sep = "\t")

gtdb.bac %>% mutate(bin_id=user_genome) %>%
  select(bin_id, classification) %>%
  bind_rows(gtdb.arc %>% mutate(bin_id=user_genome) %>%
              select(bin_id, classification)) %>%
  mutate(bin_id=gsub("bin.", "bin_", bin_id)) -> gtdb


# Genes and Contigs in gff
gff_cols <- c("seqname","source","feature","start","end","score","strand",
              "frame","attributes")
read.delim("../data/perfect_mapping/B_db_1kb.gff3",
           header=F, comment.char="#", ) -> gff
colnames(gff) <- gff_cols

# removing sequence info from the end of the gff file
gff %>%
  filter(row_number() < 1224073) %>%
  mutate(ID= gsub(";.*", "", attributes)) %>%
  mutate(ID=gsub("ID=", "", ID)) %>%
  # annotated regions that have only contig info in attributes. 
  # I don't need them
  filter(feature!= "region")-> gff


# Genes commonly engrafted in >= 3 responder
int_genes_table <- read.csv("../data/genes_FMT-Res/gene_table.txt",
                      header = T, sep = "\t")

int_genes_table %>%
select(Marker, Fig_lab) %>%
distinct() %>%
rename(ID=Marker) -> int_genes

int_genes %>%
  left_join(gff) %>%
  left_join(c_inBin %>% mutate(seqname=contig)) %>%
  mutate(bin_id= case_when(is.na(bin_id) ~ paste("NoBin"),
                           TRUE ~ paste(bin_id))) -> int_genes_inBins

int_genes_inBins %>%
  select(ID, bin_id, Fig_lab) %>%
  mutate(count= 1) %>% 
  spread(Fig_lab, count, fill = 0) %>%
  gather(Fig_lab, count, -ID, -bin_id) %>%
  filter(count > 0) %>%
  group_by(Fig_lab, bin_id) %>%
  tally() %>%
  mutate(Type= case_when(bin_id %in% MAGs ~ paste("MAG"),
                         bin_id == "NoBin" ~ paste("NoBin"),
                         TRUE ~ paste("Bin"))) %>%
 mutate(gene_num= case_when(n <= 10 ~ paste("<=10"),
                             between(n, 10, 50) ~ paste("10-50"),
                             between(n, 50, 250) ~ paste("50-250"),
                             between(n, 250, 550) ~ paste("250-550"),
                             TRUE ~ paste("other"))) -> plt_df

ggplot(plt_df, aes(reorder_within(bin_id, -n, Type), 
                   reorder_within(Fig_lab, -n, 1), fill=gene_num)) +
  geom_tile() +
  facet_grid(~Type, space = "free", scales = "free") +
   scale_fill_manual(values = c("<=10" = "#bae4bc",
                               "10-50" = "#7bccc4",
                               "50-250" = "#43a2ca",
                               "250-550" = "#0868ac"), na.value = "white") +
  theme_classic() +
  scale_x_reordered() + scale_y_reordered() +
  theme(axis.text.x = element_blank(),
        axis.title = element_blank(),
        legend.position = "none",
        strip.text.x = element_blank())
ggsave("../plots_paper/genes_ce_inBins.svg",
       width = 12, height = 3, units = "cm")


int_genes_inBins %>%
  mutate(Type= case_when(bin_id %in% MAGs ~ paste("MAG"),
                         bin_id == "NoBin" ~ paste("NoBin"),
                         TRUE ~ paste("Bin"))) %>%
  select(ID, bin_id, Type) %>% distinct() -> int_genes_inBins.stat

int_genes_inBins.stat %>%
  group_by(Type) %>%
  summarize(genes=n())

int_genes_inBins.stat %>%
  group_by(Type, bin_id) %>%
  tally() %>%
  group_by(Type) %>%
  summarize(bins=n())

```

gene signature in refseq.
The list of 1488 genes (../data/genes_FMT-Res/gene_ids.txt) extracted from
the donor B gene assembly, blasted against refseq and then clustered at 90%
and annotated via eggnog. (see gene_cluster_function script for details)


a function to parse, blastoutput against refseq:
gene length must be over 100bp

```{r}

blastn_parser <- function(table1, table2, identity, coverage) {
  # table1: .fna.fai file of genes, showing genes length, start and stop
  # using samtools faidx your_file.fna
  # table2: .blastout standard blastout of genes.fna against refseq
  # identity: pident cutoff
  # coverage: query coverage cutoff

gene_length <- read.csv(table1, sep = "\t", header = F)
gene_length %>%
  select(V1, V2) %>%
  rename(qseqid=V1,
         gene_length=V2) -> gene_length

blastout <- read.csv(table2, sep = "\t", header = F)
col_names <- c("qseqid", "sseqid", "pident", "length",
              "mismatch", "gapopen", "qstart", 
              "qend", "sstart", "send",
              "evalue", "bitscore")

colnames(blastout) <- col_names
blastout %>%
  left_join(gene_length) %>%
  # gene length must be over 100bp
  filter(gene_length >= 100 & pident >= identity) %>%
  group_by(qseqid, sseqid, gene_length) %>%
  summarise(total_length= sum(length),
            hit_count=n()) %>%
  ungroup() %>%
  mutate(query_coverage= total_length/gene_length*100) %>%
  # over 95% of gene of interest must be covered
  filter(query_coverage >= coverage)

}

```

Load the data to the parser function and visualize number of FMT-Res genes
in RefSeq genomes:

the selected cut-off for hits:
gene length >= 100bp, pident >= 90, qcov >= 90

```{r}

refseq_name <- read.csv("../data/genes_FMT-Res/references_headers.txt",
                        sep = "\t", header = F)
refseq_name %>%
  mutate(sseqid= gsub(" .*", "", V1)) %>%
  mutate(sseqid= gsub(">", "", sseqid)) %>%
  mutate(filename=gsub(".*__", "", V1)) %>%
  rename(refseq_header=V1)-> refseq_name
gtdb_refseq <- read.csv("../data/genes_FMT-Res/gtdb.bac120.summary.tsv",
                        sep = "\t") 
gtdb_refseq %>%
  select(user_genome, classification) %>%
  separate(classification, c("Kingdom", "Phylum", "Class",
                                   "Order", "Family", "Genus",
                                   "Species"), ";.__") %>%
  mutate(filename= paste(user_genome, ".fna.gz", sep = "")) %>%
  left_join(refseq_name) -> refseq_taxonomy

################################################################################

table1 <- "../data/genes_FMT-Res/gene_ids.fna.fai"
table2 <- "../data/genes_FMT-Res/gene_ids_inRefSeq.blastout"
blastn_parser(table1, table2, identity = 90, coverage = 90) %>%
  group_by(sseqid) %>%
  summarise(genes=n()) %>%
  left_join(refseq_taxonomy) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(spID= paste(Species, row_number(), sep = "_")) -> refseq_in3pts
# genomes with >= 5 genes:
refseq_in3pts %>%
  filter(genes >= 5) %>%
ggplot(aes(genes, reorder_within(spID, genes, Family), fill="CEGs")) +
  geom_bar(stat = "identity") +
  facet_grid(Family~., space = "free", scales = "free") +
  scale_x_log10() +
  scale_fill_manual(values = c("CEGs" = "#008837")) +
  scale_y_reordered() +
  theme_classic() +
  theme(strip.text.y = element_blank(),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size=5),
        axis.title = element_blank())
ggsave("../plots_paper/genes_ce_inRefSeq.svg",
       width = 6, height = 12, units = "cm")


refseq_in3pts %>%
  filter(genes >= 5) -> test


```

comparing genes across genomes. How unique are these genes?
do they get repeated in difference genomes are not?

```{r}

refseq_in3pts %>%
  ungroup %>%
  filter(genes >= 5) %>%
  select(sseqid) %>% distinct() %>% pull() -> top_genome_in3pts

blastn_parser(table1, table2, identity = 90, coverage = 90) %>%
  filter(sseqid %in% top_genome_in3pts) %>%
  group_by(sseqid, qseqid) %>%
  summarise(genes=n()) %>%
  left_join(refseq_taxonomy) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(spID= paste(Species, row_number(), sep = "_")) %>%
# visualize
ggplot(aes(qseqid, reorder_within(sseqid, genes, Species), fill="CEGs")) +
  geom_tile() +
  facet_grid(Family~Genus, space = "free", scales = "free") +
  scale_fill_manual(values = c("CEGs" = "#008837")) +
  scale_y_reordered() +
  theme_classic() +
  theme(strip.text.y = element_blank(),
        strip.text.x = element_blank(),
        legend.position = "none",
        axis.text = element_blank(),
        axis.title = element_blank())
ggsave("../plots_paper/genes_ce_inRefSeq_hm.svg",
       width = 6, height = 12, units = "cm")

```

## Metagenomic survey for CEGs in responders
Reading mapfiles for SRA studies:

```{r}

mpath="../data/genes_FMT-Res/"
Run <- read.csv(paste(mpath, "PRJNA400072_SRA.txt", sep = ""))
Diag <- read.csv(paste(mpath, "PRJNA400072_Diagnosis.txt", sep = ""))
Diag %>%
  mutate(Sample.Name= case_when(grepl("^[0-9]", 
                                      local_sample_id) ~ paste("PRISM",
                                                               local_sample_id,
                                                               sep = "_"),
                                TRUE ~ as.character(local_sample_id))) %>%
  select(Sample.Name, Diagnosis) %>%
  left_join(Run, by = "Sample.Name") -> PRJNA400072

############################################################################

#total seqeuncing depth for each sample:
SRAreads <- read.csv(paste(mpath, "PRJNA400072_sequencing_depth.txt", sep = ""),
                     sep = "\t")
PRJNA400072 %>%
  select(Run, Sample.Name, Diagnosis) %>%
  rename(SampleName=Sample.Name) %>%
  mutate(Dataset= paste("PRJNA400072")) %>%
  mutate(sample.id=Run) %>%
  left_join(SRAreads)-> SRAmapfile

table(SRAmapfile$Diagnosis)


```



```{r}
cols <- c("sample.id", "rname", "startpos", "endpos", 
          "numreads","covbases",
          "coverage", "meandepth", "meanbaseq", "meanmapq")

path=paste(mpath, "PRJNA400072", sep = "")
patt=".cover"
data.frame(sample.id = paste(dir(path, 
                                 pattern = patt))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read_tsv(
  file.path(path, .), 
            col_names = F))) %>%
         unnest() %>%
  mutate(sample.id = gsub(patt, "", sample.id)) %>%
  filter(X2 != "startpos") -> SRAcoverage
colnames(SRAcoverage) <- cols


SRAcoverage %>%
  mutate(cover_type= case_when(as.numeric(coverage) >= 100 ~ paste("Pos"),
                               TRUE ~ paste("Neg"))) %>%
  group_by(sample.id, cover_type) %>% tally() -> SRA_cegs_counts



SRAcoverage %>%
  mutate(cover_type= case_when(as.numeric(coverage) >= 100 ~ paste("Pos"),
                               TRUE ~ paste("Neg"))) %>%
  group_by(sample.id, cover_type) %>%
  summarise(reads= sum(as.numeric(numreads))) %>%
  left_join(SRAreads) %>%
  mutate(relativeAbund= 
           as.numeric(reads) / as.numeric(Total_reads) * 100)-> SRA_cegs_abund

```

```{r}


SRAmapfile %>%
  left_join(SRA_cegs_counts) %>%
  filter(cover_type== "Pos") %>%
  mutate(Diagnosis= factor(Diagnosis, 
                           levels = c("Control", "UC", "CD"))) -> tbl1

model=lm( tbl1$n ~ tbl1$Diagnosis )
ANOVA=aov(model)
TUKEY <- TukeyHSD(x=ANOVA, 'tbl1$Diagnosis', conf.level=0.95)
TUKEY
plot(TUKEY , las=1 , col="brown")

ggplot(tbl1, aes(Diagnosis, n)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Diagnosis),size=2,shape=21, 
             position = position_dodge(0.2)) + 
  scale_fill_manual(values = c("Control" = "#1a9641",
                                "UC" = "#d7191c",
                                "CD" = "#d7191c")) +
  theme_classic() +
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")
  #ylab("# of unique CEGs / sample")
ggsave("../plots_paper/genes_ce_inPublicMeta_uniq.svg",
       width = 5, height = 5, units = "cm")

############################################################################

SRAmapfile %>%
  left_join(SRA_cegs_abund) %>%
  filter(cover_type== "Pos") %>%
  mutate(Diagnosis= factor(Diagnosis, 
                           levels = c("Control", "UC", "CD"))) -> tbl2

model=lm( tbl2$relativeAbund ~ tbl2$Diagnosis )
ANOVA=aov(model)
TUKEY <- TukeyHSD(x=ANOVA, 'tbl2$Diagnosis', conf.level=0.95)
TUKEY
plot(TUKEY , las=1 , col="brown")

ggplot(tbl2, aes(Diagnosis, relativeAbund)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Diagnosis),size=2,shape=21, 
             position = position_dodge(0.2)) + 
  scale_fill_manual(values = c("Control" = "#1a9641",
                                "UC" = "#d7191c",
                                "CD" = "#d7191c")) +
  theme_classic() +
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")
  #ylab("% of CEGs reads / sample")
ggsave("../plots_paper/genes_ce_inPublicMeta_abund.svg",
       width = 5, height = 5, units = "cm")



```

