---
title: "Metagenomic_reads"
author: "Sharok"
date: "4/28/2023"
output: html_document
---

```{r}

library(tidyverse)
library(tidytext)

```

```{r}

mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

```

## Marker-level analysis
Reading marker abundance files, parse them into one
```{r}
path="../data/Metaphlan4/marker_abund"
#outpath="metaphlan4"
data.frame(sample.id = paste(dir(path,
                                 pattern = "*_abund_norm.txt"))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read.csv(file.path(path, .),
                                                 sep = "\t", header = F,
                                                 comment.char = "#"))) %>%
unnest() %>%
mutate(sample.id= gsub(".marker_abund_norm.txt", "", sample.id)) %>%
  spread(sample.id, V2) %>%
  rename(Marker="V1") %>%
  mutate(temp=Marker) %>%
  separate(temp, c("sp_marker", "temp1", "temp2"), sep = "__") %>%
  mutate(st_marker= case_when(!is.na(temp2) ~ paste(temp1, temp2, sep = "__"),
                              TRUE ~ paste(temp1))) %>%
  select(sp_marker, st_marker, Marker, DonorB_D_2013:PMCL883) -> marker_lvl


```


```{r}

################################################################################
# a function to detect common engraftment based on strains (metaphlan markers)
################################################################################
stCE <- function(marker_lvl, mapfile, strains, cutoff=cutoff, ...){
  # convert marker-lvl to long format, make Study_ID variable (in mapfile),
  # and select only donor's strains
  marker_lvl %>%
  filter(Marker %in% strains) %>%
  select( -sp_marker, -st_marker) %>%
  gather(sample, abundance, -Marker) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_markerlvl
  # complete list of engrafted
  long_markerlvl %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(Marker, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -Marker) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i),
                             E.strains = length(Emark.1[[i]]$Marker))
  }
  do.call(rbind.data.frame, Emark.2)
}
################################################################################
################################################################################

```


commonly engraftment of markers:
```{r}

marker_lvl %>%
  gather(sample, abundance, -sp_marker, -st_marker, -Marker) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) -> donorB_markers

##### seting engraftment cut-off
donorB_markers %>%
  group_by(sample) %>%
  # minimum detection in each donor B sample
  slice(which.min(abundance)) -> min_B_markers
  # average of minimum detections-> test
cutoff= mean(min_B_markers$abundance)
######

# selecting only those donor B markers that meet below cutoff:
# > average of minimum detections
donorB_markers %>%
  filter(as.numeric(abundance) > cutoff) -> donorB_markers

donorB_markers %>%
  select(Marker) %>% distinct() %>%
  pull() -> B_markers



######### common engraftment for FMT ###############################
stCE(marker_lvl, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "FMT") -> commonE_FMT
######### common engraftment for Placebo ############################
stCE(marker_lvl, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "Placebo") -> commonE_Placebo
######### common engraftment for Responder #########################
stCE(marker_lvl, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "FMT" & Remission == "Res") -> commonE_FMTRes
######### common engraftment for Non-responder #####################
stCE(marker_lvl, mapfile, strains = B_markers, cutoff = cutoff,
     Treatment == "FMT" & Remission == "NoRes") -> commonE_FMTNoRes

######################################################
########### visualizing
######################################################

cols1 <- c("FMT" = "#008837",
           "Placebo" = "#bdbdbd",
           "Responder" = "#0571b0",
           "Non-Responder" = "#ca0020")

commonE_FMT %>% 
  mutate(Type=paste("FMT"),
          Samples= "All") %>%
  bind_rows(commonE_Placebo %>% 
             mutate(Type=paste("Placebo"),
                    Samples="All")) %>%
  bind_rows(commonE_FMTRes %>%
              mutate(Type="Responder",
                     Samples="onlyFMT")) %>%
  bind_rows(commonE_FMTNoRes %>%
              mutate(Type="Non-Responder",
                     Samples="onlyFMT")) -> tbl

tbl$Type <- factor(tbl$Type, levels = c("FMT", "Placebo", 
                                        "Responder", "Non-Responder"))
ggplot(tbl, aes(as.numeric(Npts), E.strains, color=Type)) +
  geom_line(aes(group=Type))+
  #facet_grid(~Samples, space = "free") +
  scale_y_log10() +
  scale_x_continuous(breaks = c(1, 2, 3, 4, 5, 6, 7,8,9,10,11,12)) +
  scale_color_manual(values = cols1) +
  theme_classic() +
  theme(legend.title = element_blank(),
        legend.position = c("0.8", "0.8")) +
  xlab("Increasing # of patients") +
  ylab("# of engrafted Strains")
ggsave("../figs/CESt.png",
       width = 10, height = 8, units = "cm")

```

Reading metaphlan4 default output (merged)
relative abundance
```{r}

org <- read.csv("../data/Metaphlan4/merged_metaphlan4.txt", 
                sep = "\t", header = F, 
                comment.char = "#")
org %>%
  filter(V1 != "clade_name") %>%
  # selecting only at strain level- contains all lineage- avoid duplicates
  filter(grepl("t__S", V1)) -> metaphlan
colnames(metaphlan) <- gsub(".mp.profile_sp", "", org[1,])

metaphlan %>%
  mutate(temp=clade_name) %>%
  separate(temp, c("Kingdom", "Phylum", "Class", 
                       "Order", "Family", "Genus", 
                       "Species", "marker"), sep = "[|]") %>%
  select(clade_name,Kingdom:marker) %>%
  mutate(sp_marker= gsub("t__", "", marker)) %>%
  mutate(sp_marker=gsub("_group", "", sp_marker))-> lineage


```


```{r}

################################################################################
# a function to detect common engraftment based on species (metaphlan)
################################################################################
spCE <- function(metaphlan, mapfile, species, cutoff=cutoff, ...){
  # convert metaphlan to long format, make Study_ID variable (in mapfile),
  # and select only donor's species
  metaphlan %>%
  filter(clade_name %in% species) %>%
  gather(sample, abundance, -clade_name) %>%
  mutate(Study_ID = case_when(
    str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
    TRUE ~ paste(sample))) -> long_metaphlan
  # complete list of engrafted
  long_metaphlan %>%
    left_join(mapfile) %>%
    filter(...) %>%
    spread(Timepoint, abundance, fill = 0) %>%
    mutate(model= case_when(WK0 <= 0 & WK6 > cutoff ~ paste("Engraft"),
                          TRUE ~ paste("other"))) %>%
    filter(model == "Engraft") %>%
    select(clade_name, Study_ID, WK6) %>%
    spread(Study_ID, WK6, fill = 0) -> engraft
  # number of samples in input table
  engraft %>%
    gather(sample, abund, -clade_name) %>% 
    pull(sample) %>% n_distinct() -> number_of_samples
  # for loop for engraftment across increasing number of patients
  Emark.1 <- list()
  Emark.2 <- list()
  for (i in 1:number_of_samples) {
  Emark.1[[i]] <- engraft[rowSums(engraft[,2:ncol(engraft)] > 0) >= i, ]
  Emark.2[[i]] <- data_frame(Npts = paste(i),
                             E.species = length(Emark.1[[i]]$clade_name))
  }
  do.call(rbind.data.frame, Emark.2)
}
################################################################################
################################################################################

```


commonly engraftment of species (default metaphlan):
```{r}
mapfile %>% select(Study_ID, Treatment, Timepoint) -> map

metaphlan %>%
  gather(sample, abundance, -clade_name) %>%
  #filtering donor B markers
  filter(str_detect(sample, "DonorB_D_")) %>%
  #any marker with abundance > 0
  #markers that were present in at least one donor B sample
  filter(!is.na(abundance)) %>%
  filter(as.numeric(abundance) > 0) -> donorB_species

##### seting engraftment cut-off
donorB_species %>%
  group_by(sample) %>%
  # minimum detection in each donor B sample
  slice(which.min(abundance)) -> min_B_species
  # average of minimum detections-> test
cutoff= mean(as.numeric(min_B_species$abundance))
######

# working with only donor B species that meeet cutoff
# > average of minimum detections
donorB_species %>%
  filter(as.numeric(abundance) > cutoff) -> donorB_species

  
donorB_species %>%
  select(clade_name) %>% distinct() %>%
  pull() -> B_species
  

######### common engraftment for FMT ################################
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment == "FMT") -> commonE_FMT

######### common engraftment for Placebo ############################
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, Treatment == "Placebo") -> commonE_Placebo

######### common engraftment for Responder #########################
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, 
     Treatment == "FMT" & Remission == "Res") -> commonE_FMTRes
######### common engraftment for Non-responder #####################
spCE(metaphlan, mapfile, species = B_species,
     cutoff = cutoff, 
     Treatment == "FMT" & Remission == "NoRes") -> commonE_FMTNoRes


######################################################
########### visualizing
######################################################

cols1 <- c("FMT" = "#008837",
           "Placebo" = "#bdbdbd",
           "Responder" = "#0571b0",
           "Non-Responder" = "#ca0020")

commonE_FMT %>% 
  mutate(Type=paste("FMT"),
          Samples= "All") %>%
  bind_rows(commonE_Placebo %>% 
             mutate(Type=paste("Placebo"),
                    Samples="All")) %>%
  bind_rows(commonE_FMTRes %>%
              mutate(Type="Responder",
                     Samples="onlyFMT")) %>%
  bind_rows(commonE_FMTNoRes %>%
              mutate(Type="Non-Responder",
                     Samples="onlyFMT")) -> tbl
tbl$Type <- factor(tbl$Type, levels = c("FMT", "Placebo", 
                                        "Responder", "Non-Responder"))
ggplot(tbl, aes(as.numeric(Npts), E.species, color=Type)) +
  geom_line(aes(group=Type))+
  #facet_grid(~Samples, space = "free") +
  scale_y_log10() +
  scale_x_continuous(breaks = c(1, 2, 3, 4, 5, 6, 7,8,9,10,11,12)) +
  scale_color_manual(values = cols1) +
  theme_classic() +
  theme(legend.title = element_blank(),
        legend.position = c("0.8", "0.8")) +
  xlab("Increasing # of patients") +
  ylab("# of engrafted Species")
ggsave("../figs/CESp.png",
       width = 10, height = 8, units = "cm")

```

## profile microbial taxa

```{r}

family_col= c(
"f__Lachnospiraceae" = "#6a3d9a",
"f__Ruminococcaceae" = "#33a02c",
"f__Bacteroidaceae" = "#1f78b4",
"f__Bacteroidales_unclassified" = "#ff7f00",
"f__Prevotellaceae" = "#b2df8a",
"f__Veillonellaceae" = "#fdbf6f",
"f__Bifidobacteriaceae" = "#ffff99",
"f__Clostridia_unclassified" =  "#a6cee3",
"f__Enterobacteriaceae" = "#e31a1c",
"f__Erysipelotrichaceae" = "#f781bf",
"f__Streptococcaceae" = "#cab2d6",
"f__Clostridiaceae" = "#fb9a99",
"f__Rikenellaceae" = "#b15928",
"f__Coriobacteriaceae" = "#8dd3c7",
"f__Bacilli_unclassified" = "#fb8072",
"f__Enterococcaceae" = "#fccde5",
"f__Desulfovibrionaceae" = "#80b1d3",
"f__Comamonadaceae" = "#b3de69",
"f__Peptostreptococcaceae" = "#ccebc5",
"f__Lactobacillaceae" = "#dfc27d",
"other" = "#808080"
)


```


```{r}

mapfile %>%
  mutate(sample= gsub(".R1.fastq.gz", "", File)) %>%
  mutate(Timepoint= case_when(grepl("DonorB_D_", sample) ~ paste("DMG"),
                              grepl("DonorB_P", sample) ~ paste("CEMG"),
                              TRUE ~ paste(Timepoint))) %>%
  mutate(Treatment= case_when(grepl("DonorB_D_", sample) ~ paste("DMG"),
                              grepl("DonorB_P", sample) ~ paste("CEMG"),
                              TRUE ~ paste(Treatment)))-> metaphlan_map
metaphlan %>%
  gather(sample, abundance, -clade_name) %>%
  left_join(metaphlan_map) %>%
  separate(clade_name, c("Kingdom", "Phylum", "Class", "Order", "Family", 
                         "Genus", "Species", "Strain"), sep = "[|]") %>%
  mutate(abundance= as.numeric(abundance)) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> tbl


################################
#select top 20 families
tbl %>%
  count(Family, wt = abundance) %>%
  filter(!is.na(Family)) %>%
  arrange(desc(n)) %>%
  pull(Family) -> tfam
  tfam20 <- tfam[1:20]
  
################################
# edit a new family column
tbl %>%
  mutate(Family_lab= case_when(Family %in% tfam20 ~ paste(Family),
                               TRUE ~ paste("other"))) -> tbl
# sum the abundance for each class, across all IDs, & sort the result
  tbl%>%
  count(Family_lab, wt = abundance) %>%
  arrange(desc(n)) %>%
  pull(Family_lab) -> sort.family
# get ID order, sorted by each ID's abundance in the most abundant class
  tbl %>%
  mutate(Fig_lab= case_when(is.na(Fig_lab) ~ paste(Study_ID),
                            TRUE ~ paste(Fig_lab))) %>%
  filter(Family_lab == sort.family[1]) %>%
  arrange(desc(abundance)) %>% distinct(Fig_lab) %>%
  pull(Fig_lab) -> ID.order

tbl$Fig_lab <- factor(tbl$Fig_lab, levels = ID.order)
tbl$Family_lab <- factor(tbl$Family_lab, levels = rev(sort.family))
tbl$Timepoint <- factor(tbl$Timepoint, levels = c("WK0", "WK6", "DMG", "CEMG"))
tbl$Treatment <- factor(tbl$Treatment, levels = c("Placebo", "FMT", "DMG", "CEMG"))


p1 <- tbl %>%
  filter(Treatment %in% c("Placebo", "FMT")) %>%
ggplot(aes(abundance, Fig_lab, fill=Family_lab, color=Family_lab)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = family_col) +
  scale_color_manual(values = family_col) +
  facet_grid(Treatment~Timepoint, scales = "free", space = "free") +
  theme_classic() +
  theme(legend.position = "none", 
        axis.title = element_blank())

p2 <- tbl %>%
  filter(Treatment %in% c("CEMG", "DMG")) %>%
ggplot(aes(abundance, sample, fill=Family_lab, color=Family_lab)) +
  geom_bar(stat = "identity") +
  facet_grid(Timepoint~., scales = "free", space = "free") +
  scale_fill_manual(values = family_col) +
  scale_color_manual(values = family_col) +
  theme_classic() +
  theme(legend.position = "left", 
        legend.title = element_blank(),
        axis.title = element_blank()) +
  guides(fill=guide_legend(ncol =1))

cowplot::plot_grid(p1, p2, ncol = 2)
ggsave("../figs/RA_profile.png",
       width = 30, height = 15, units = "cm")

```

## Shannon diversity

```{r}

shannon <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                                "merged_metaphlan4_shannon.tsv", 
                      sep = ""), sep = "\t")

shannon %>%
  rownames_to_column("ID") %>%
  mutate(sample= gsub(".mp.profile_sp", "", ID)) %>%
  left_join(metaphlan_map) %>%
  # remove plate pools
  filter(!Timepoint %in% c("DMG","CEMG")) -> tbl

ggplot(tbl, aes(x = Timepoint, y = diversity_shannon)) +
  geom_boxplot(outlier.color = NA) +
  geom_line(aes(group=Fig_lab), position = position_dodge(0.2), alpha = 0.2) +
  #geom_jitter(aes(color = Timepoint), height = 0, width = .2) +
  geom_point(aes(fill=Remission,
                 group=Fig_lab),size=2,shape=21, 
             position = position_dodge(0.2)) + 
   facet_grid(~Treatment, scales = "free") +
   scale_fill_manual(values = c("#0571b0", "#ca0020"),
                       breaks = c("Res", "NoRes")) +
    ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Shannon")
ggsave("../figs/Shannon_div.png",
       width = 6, height = 6, units = "cm")

# We happen to have lower diversity in patients on placebo; this observed both
# in baseline and post-FMT
# other than treatment group, the difference between timepoint is not sig diff
# in FMT compared to placebo
mod = lm(data=tbl, diversity_shannon  ~ Timepoint * Treatment)
anova(mod)
em = emmeans::emmeans(mod, "Timepoint", by = "Treatment")
as.data.frame(summary(em))

```

## Bray-curtis

```{r}

braycurtis <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                       "merged_metaphlan4_bray-curtis.tsv", 
                      sep = ""), sep = "\t")
colnames(braycurtis) <- gsub(".mp.profile_sp", "", colnames(braycurtis))
rownames(braycurtis) <- gsub(".mp.profile_sp", "", rownames(braycurtis))

braycurtis %>%
rownames_to_column("sample_1") %>%
  gather(sample_2, value, -sample_1) %>%
  filter(as.character(sample_1) != as.character(sample_2)) %>%
  mutate_if(is.factor, as.character) -> dis.tbl

# two mapfile for each variable in matrix
metaphlan_map -> V1.tbl
colnames(V1.tbl) <- paste(colnames(V1.tbl), "1", sep = "_")
metaphlan_map-> V2.tbl
colnames(V2.tbl) <- paste(colnames(V2.tbl), "2", sep = "_")

#merging all info together
dis.tbl %>%
  left_join(V1.tbl) %>%
  left_join(V2.tbl) %>%
  # remove plate pools
  filter(Treatment_1 != "CEMG" & Treatment_2 != "CEMG") %>%
  # compare samples within individuals
  filter(Fig_lab_1 == Fig_lab_2 | sample_1 == "DonorB_D_2013") %>%
  select(sample_1, sample_2, value, Fig_lab_1, Fig_lab_2,
         Timepoint_1, Timepoint_2, Treatment_1, Treatment_2,
         Remission_2) %>%
  mutate(Time_dis= case_when(Timepoint_1 == "DMG" & 
                               Timepoint_2 == "DMG" ~ paste("Donor"),
                             Timepoint_1 == "DMG" ~ paste("Donor", Timepoint_2,
                                                          sep = "-"),
                             TRUE ~ paste(Timepoint_1, Timepoint_2, 
                                          sep = "-"))) %>%
  # WE ARE SHOWING WK0<>WK6 which is same as below, so remove:
  filter(Time_dis != "WK6-WK0") %>%
  mutate(Remission_2= case_when(is.na(Remission_2) ~ paste("Donor"),
                                TRUE ~ paste(Remission_2)))-> tbl


ggplot(tbl, aes(Treatment_2, value)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Remission_2,
                 group=Fig_lab_1),size=2,shape=21, 
             position = position_dodge(0.2)) +
  facet_grid(~Time_dis, space = "free", scales = "free") +
  scale_fill_manual(values = c("#0571b0", "#ca0020", "#008837"),
                       breaks = c("Res", "NoRes", "Donor")) +
  ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Bray-Curtis")
ggsave("../figs/Bray-Curtis_div.png",
       width = 10, height = 6, units = "cm")

```

## Aitchison

```{r}

aitchison <- read.csv(paste("../data/Metaphlan4/diversity_analysis/",
                       "merged_metaphlan4_aitchison.tsv", 
                      sep = ""), sep = "\t")
colnames(aitchison) <- gsub(".mp.profile_sp", "", colnames(aitchison))
rownames(aitchison) <- gsub(".mp.profile_sp", "", rownames(aitchison))

aitchison %>%
rownames_to_column("sample_1") %>%
  gather(sample_2, value, -sample_1) %>%
  filter(as.character(sample_1) != as.character(sample_2)) %>%
  mutate_if(is.factor, as.character) -> dis.tbl

# two mapfile for each variable in matrix
metaphlan_map -> V1.tbl
colnames(V1.tbl) <- paste(colnames(V1.tbl), "1", sep = "_")
metaphlan_map-> V2.tbl
colnames(V2.tbl) <- paste(colnames(V2.tbl), "2", sep = "_")

#merging all info together
dis.tbl %>%
  left_join(V1.tbl) %>%
  left_join(V2.tbl) %>%
  # remove plate pools
  filter(Treatment_1 != "CEMG" & Treatment_2 != "CEMG") %>%
  # compare samples within individuals
  filter(Fig_lab_1 == Fig_lab_2 | sample_1 == "DonorB_D_2013") %>%
  select(sample_1, sample_2, value, Fig_lab_1, Fig_lab_2,
         Timepoint_1, Timepoint_2, Treatment_1, Treatment_2,
         Remission_2) %>%
  mutate(Time_dis= case_when(Timepoint_1 == "DMG" & 
                               Timepoint_2 == "DMG" ~ paste("Donor"),
                             Timepoint_1 == "DMG" ~ paste("Donor", Timepoint_2,
                                                          sep = "-"),
                             TRUE ~ paste(Timepoint_1, Timepoint_2, 
                                          sep = "-"))) %>%
  # WE ARE SHOWING WK0<>WK6 which is same as below, so remove:
  filter(Time_dis != "WK6-WK0") %>%
  mutate(Remission_2= case_when(is.na(Remission_2) ~ paste("Donor"),
                                TRUE ~ paste(Remission_2)))-> tbl

ggplot(tbl, aes(Treatment_2, value)) +
  geom_boxplot(outlier.color = NA) +
  geom_point(aes(fill=Remission_2,
                 group=Fig_lab_1),size=2,shape=21, 
             position = position_dodge(0.2)) +
  facet_grid(~Time_dis, space = "free", scales = "free") +
  scale_fill_manual(values = c("#0571b0", "#ca0020", "#008837"),
                       breaks = c("Res", "NoRes", "Donor")) +
  ggplot2::theme_classic() +
    theme(legend.title = element_blank(),
        text = element_text(size = 10),
        legend.position = "none", axis.title.x = element_blank()) +
  ylab("Aitchison")
ggsave("../figs/Aitchison_div.png",
       width = 10, height = 6, units = "cm")

```



