---
title: "metagenomics_genes"
author: "Sharok"
date: "2023-09-13"
output: html_document
---

```{r}

library(tidyverse)
library(tidytext)

```


```{r}

mapfile <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")

```

## reading gene coverage

```{r}

heads_cover <- c("sample.id",
                 "gene_id", "startpos", "endpos", 
                 "numreads", "covbases", "coverage", "meandepth",
                 "meanbaseq", "meanmapq",
                 "ReadSampling", "Quality")

path="../data/perfect_mapping/B_db_bakta_cover_subreads"
patt="_perfect*.cover"
data.frame(sample.id = paste(dir(path,
                                 pattern = patt))) %>%
# read each file from the directory (current path) and then unnest
mutate(file_contents = map(sample.id, ~ read.csv(file.path(path, .),
                                                 sep = "\t", header = F,
                                                 comment.char = "#"))) %>%
unnest() %>%
mutate(sample.id = gsub("_perfect_BDB.cover", "", sample.id)) %>%
mutate(sample.id = gsub("sub_", "", sample.id))-> marker_lvl_RA
colnames(marker_lvl_RA) <- heads_cover

marker_lvl_RA %>%
  select(sample.id, gene_id, coverage, meandepth) %>%
  mutate(sample.id=gsub("_perfect.cover", "", sample.id)) %>%
  # Meandepth of genes should be >= 1
  filter(meandepth >= 1) %>% select(-meandepth) %>%
  rename(Marker=gene_id) %>%
  spread(sample.id, coverage, fill=0) -> marker_lvl_RA
  
cutoff = 90

```


## Functions

```{r}
# Make the marker data frame long
mark_to_long = function(marker_lvl) {
  # This can be done once. It doesn't need to be repeated each time
  # convert marker-lvl to long format and make StudyID variable to match mapfile
  # and select only donor's strains
  marker_lvl %>%
    gather(sample, abundance, -Marker) %>%
    mutate(Study_ID = case_when(
      str_detect(sample,"PMCL") ~ paste(gsub("_.*", "",sample)),
      TRUE ~ paste(sample))) -> long_markerlvl
  return(long_markerlvl)
}

################################################################################
# a function to detect common engraftment based on strains (metaphlan markers)
################################################################################
get_engraft <- function(long_markerlvl, mapfile, cutoff=cutoff, ...){
  # complete list of engrafted
  engraft <- (long_markerlvl 
              %>% left_join(mapfile) 
              %>% filter(...)
              %>% spread(Timepoint, abundance, fill = 0) 
              %>% filter(WK0 < 5 & WK6 >= cutoff) 
              %>% select(Marker, Study_ID, WK6) 
              %>% spread(Study_ID, WK6, fill = 0) 
              %>% gather(Study_ID, WK6, -Marker)
              %>% left_join(mapfile))
  return(engraft)
}

count_engraft <- function(engraft, group_cols){
  # Count how many markers engrafted in how many patients
  g1 = group_cols
  g2 = c('Marker', group_cols)
  ct_df = (engraft
           %>% group_by(.[,g2])
           %>% summarize(n = sum(WK6 > 0))
           %>% pivot_wider(names_from = {{ g1 }},
                           values_from = n)
           %>% ungroup()
           %>% mutate(Total = rowSums(across(2:ncol(.)))))
  # Uncomment the next bit if you want the cumulative sums, but it adds 0.8s
  # to the run time!
  # ct_df = (ct_df
  #          %>% mutate(n = rev(cumsum(rev(n))))) 
  return(ct_df)
  
}
################################################################################


```


### Observed AUC: FMT vs Placebo

```{r}

# Remove unnecessary columns from the mapfile to speed up the loop (saves 10s per 
# loop)
map_mod = (mapfile
           %>% select(Study_ID, Timepoint, Treatment, Fig_lab))

# Converting the markers to long format only needs to be done once, saves 5s per
# loop
markers_long = mark_to_long(marker_lvl_RA)

cts = (markers_long
       %>% get_engraft(map_mod, cutoff, 
                       Treatment %in% c('FMT', 'Placebo'))
       %>% count_engraft(group_cols = 'Treatment')
       %>% mutate(Uniqueness = 
                    case_when(FMT == 0 & Placebo == 0 ~ 'zero',
                              FMT == 0 ~ 'Placebo',
                              Placebo == 0 ~ 'FMT',
                              TRUE ~ 'Both'),
                  it = 'observed'))
npts_tx = rbind(
  (cts
   %>% count(FMT)
   %>% rename(Npts = FMT)
   %>% mutate(Treatment = 'FMT')),
  (cts
   %>% count(Placebo)
   %>% rename(Npts = Placebo)
   %>% mutate(Treatment = 'Placebo')))

fmt_fn = with(filter(npts_tx, Treatment == 'FMT'),
              approxfun(Npts, n))
plc_fn = with(filter(npts_tx, Treatment == 'Placebo'),
               approxfun(Npts, n))
fmt_auc = with(filter(npts_tx, Treatment == 'FMT'),
                   integrate(fmt_fn, 1, max(Npts))$value)
plc_auc = with(filter(npts_tx, Treatment == 'Placebo'),
                   integrate(plc_fn, 1, max(Npts))$value)
obs_d_auc = fmt_auc - plc_auc

obs_d_auc

```

### Get observed AUC p-value: FMT vs Placebo
loading permuated data "auc_null"

```{r}

# load data:
auc_null<- read.csv("../permut_data/genes_auc_null.txt",
                    sep = "\t")

# Get the observed FMT values
ggplot(data.frame(auc_null), aes(x = deltaAUC)) +
    geom_density() +
    geom_vline(xintercept = obs_d_auc) +
  theme_classic()

# Place that difference on the null distribution of differences and get its p-value
quant_fun = ecdf(auc_null[,1])
auc_pval = 1 - quant_fun(obs_d_auc)
round(auc_pval, 2)

# merging the results together:
plt_df = filter(npts_tx, Npts > 0)
ribdat = plt_df %>% spread(Treatment, n, fill = 0)
  
cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999")

f1 <- ggplot(plt_df, aes(Npts, n, color=Treatment)) +
  geom_point() + geom_line() +
  geom_ribbon(data = ribdat, aes(ymin=Placebo,ymax=FMT, x = Npts),
              linetype=2, alpha=0.1, inherit.aes = FALSE)+
  annotate("text", x=9, y=100000, size=2,
          label= paste("Permutation test,\n p=", auc_pval, sep = " ")) +
  scale_x_continuous(breaks = seq(0, max(plt_df$Npts), 1)) +
  scale_color_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none",
        legend.title = element_blank(),
        axis.title = element_blank())
```

## Uniqueness
test whether the engrafted markers within FMT and Placebo are shared between
these groups or unique to each treatment

```{r}

plt2_df = (cts
           %>% count(Total, Uniqueness)
           %>% group_by(Total))


cols <- c("FMT" = "#1b7837",
          "Placebo" = "#999999",
          "Both" = "#fc8d59")

f2 <- ggplot(plt2_df, aes(x = Total, y = n, fill = Uniqueness)) +
  geom_bar(stat = 'identity') +
  scale_fill_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_blank()) 

```

### Observed AUC: Res vs No-Res

```{r}

# Remove unnecessary columns from the mapfile to speed up the loop (saves 10s per 
# loop)
map_mod = (mapfile
           %>% filter(Treatment == 'FMT')
           %>% select(Study_ID, Timepoint, Remission, Fig_lab))

fmt_cts = (markers_long
       %>% get_engraft(map_mod, cutoff, 
                       Remission %in% c('Res', 'NoRes'))
       %>% count_engraft(group_cols = 'Remission')
       %>% mutate(Uniqueness = 
                    case_when(Res == 0 & NoRes == 0 ~ 'zero',
                              Res == 0 ~ 'NoRes',
                              NoRes == 0 ~ 'Res',
                              TRUE ~ 'Both'),
                  it = 'observed'))
npts_fmt = rbind(
  (fmt_cts
   %>% count(Res)
   %>% rename(Npts = Res)
   %>% mutate(Remission = 'Res')),
  (fmt_cts
   %>% count(NoRes)
   %>% rename(Npts = NoRes)
   %>% mutate(Remission = 'NoRes')))

res_fn = with(filter(npts_fmt, Remission == 'Res'),
              approxfun(Npts, n))
nrs_fn = with(filter(npts_fmt, Remission == 'NoRes'),
               approxfun(Npts, n))
res_auc = with(filter(npts_fmt, Remission == 'Res'),
               sum(integrate(res_fn, 0, 1)$value,
                   integrate(res_fn, 1, max(Npts))$value))
nrs_auc = with(filter(npts_fmt, Remission == 'NoRes'),
               sum(integrate(nrs_fn, 0, 1)$value,
                   integrate(nrs_fn, 1, max(Npts))$value))
rm_d_auc = res_auc - nrs_auc

```

### Get observed AUC p-value: Res vs No-Res
loading permutated data "rm_auc_null"

```{r}

# load data:
rm_auc_null <- read.csv("../permut_data/genes_remission_auc_null.txt",
                    sep = "\t")

# Get the observed FMT values
ggplot(data.frame(rm_auc_null), aes(x = deltaAUC)) +
    geom_density() +
    geom_vline(xintercept = rm_d_auc) +
  theme_classic()

# Place that difference on the null distribution of differences and get its p-value
quant_fun = ecdf(rm_auc_null[,1])
auc_pval = 1 - quant_fun(rm_d_auc)
round(auc_pval, 2)


# merging the results together:
plt_df = filter(npts_fmt, Npts > 0)
ribdat = plt_df %>% spread(Remission, n, fill = 0)


cols <- c("Res" = "#2166ac",
          "NoRes" = "#b2182b")
  
f3 <- ggplot(plt_df, aes(Npts, n, color=Remission)) +
  geom_point() + geom_line() +
  geom_ribbon(data = ribdat, aes(ymin=NoRes,ymax=Res, x = Npts),
              linetype=2, alpha=0.1, inherit.aes = FALSE)+
  scale_x_continuous(breaks = seq(0, max(plt_df$Npts), 1)) +
  annotate("text", x=5, y=100000, size=2,
           label= paste("Permutation test,\n p=", auc_pval, sep = " ")) +
  scale_color_manual(values = cols) +
  theme_classic() +
  theme(legend.position = "none",
        axis.title = element_blank())

```

### saving figure
```{r}

cowplot::plot_grid(f1, f2, f3, ncol = 1, hjust = T, vjust = T)
ggsave("../figs/genes_CEg.png",
       width = 7, height = 12, units = "cm")

```

## common engraftment for >=3 pts

```{r}

fmt_cts %>%
  filter(Total >= 3) %>%
  select(Marker, Total, Uniqueness) %>%
  rename(Total_Remi=Total,
         Uniqueness_Remi=Uniqueness)-> of_int_fmt
of_int_fmt %>%
  pull(Marker) -> of_int_mark

cts %>%
  filter(Marker %in% of_int_mark) %>%
  select(Marker, Total, Uniqueness) %>%
  rename(Total_Treat=Total,
         Uniqueness_Treat=Uniqueness) -> of_int


#I need to use cts and fmt_cts to campare Res aND NoRes in FMT and Placebo 

###### based on treatment: placebo and FMT
map_mod = (mapfile
           %>% select(Study_ID, Timepoint, Treatment, Fig_lab))
get_engraft(markers_long, map_mod, cutoff, 
            Treatment %in% c('FMT', 'Placebo')) %>%
  ungroup() %>%
  filter(Marker %in% of_int_mark) %>%
  group_by(Marker, Fig_lab, Treatment) %>%
  summarize(n= sum(WK6 > 0)) %>%
  ungroup() %>%
  left_join(of_int) %>% 
  left_join(of_int_fmt) %>%
  mutate(cat=paste(Uniqueness_Treat, Uniqueness_Remi, sep = "")) %>%
  mutate(color= case_when(n > 0 & cat == "BothBoth" ~ paste("Placebo-overlap"),
                          n > 0 & cat == "BothRes" ~ paste("Placebo-overlap"),
                          n > 0 & cat == "BothNoRes" ~ paste("Placebo-overlap"),
                          n > 0 & cat == "FMTBoth" ~ paste("FMT-overlap"),
                          n > 0 & cat == "FMTRes" ~ paste("Res"),
                          n > 0 & cat == "FMTNoRes" ~ paste("NoRes"),
                          TRUE ~ paste("none"))) %>%
  left_join(mapfile %>% select(Fig_lab, Remission) %>% distinct()) %>%
  mutate(Fig_lab= case_when(Remission == "Res" ~ paste("*", Fig_lab, sep = ""),
                            TRUE ~ paste(Fig_lab))) -> plt_df

cols <- c("Res" =  "#2166ac",
          "NoRes" = "#b2182b",
          "Placebo-overlap" = "#fa5305", #"#fc8d59",
          "FMT-overlap" = "#6A3F6C")

ggplot(plt_df, aes(reorder_within(Marker, -Total_Remi, 1), Fig_lab, fill=color)) +
  geom_tile() +
  facet_grid(Treatment~cat, space = "free", scales = "free") +
  scale_fill_manual(values = cols, na.value = NA) +
  scale_x_reordered() +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.title = element_blank(),
        strip.text.x = element_blank(),
        legend.position = "none")
ggsave("../figs/genes_CEm_3pts.png",
       width = 14, height = 8, units = "cm")

plt_df %>%
  distinct() %>%
  filter(cat == "FMTRes" & n > 0) %>%
  write.table("../data/genes_of_interest/CEGs_3pts_uniquetoFMT-Res.txt",
              row.names = F, quote = F, sep = "\t", col.names = F)

plt_df %>%
  select(Marker, n, color) %>%
  filter(n > 0) %>%
  distinct() %>%
  group_by(color) %>%
  summarise(n=n()) %>%
  mutate(total=sum(n),
         relative=n/sum(n)*100) %>%
  mutate(percent=round(relative, 1))

```


