Data Objects and Descriptions
=============================

The purpose of this document is to examine and describe every data object
associated with the UCFMT1 analysis. That way, when something changes we are
able to detect the differences. It is probably a good idea to move this
eventually into a `testthat` framework, but this is where we are for now.

Setup
-----


```{r}
library(tidyverse)
source('paperfig_functions.R')
```


Input Data
----------

This is files that have been generated by previous scripts and aren't changing.
This includes the mapfiles and marker_lvl files.

### Mapfiles

```{r}
mapfile_16s <- read.csv("../data/UCFMT1_16Smap.txt")
dim(mapfile_16s)
table(select(mapfile_16s, Treatment, Remission))
mapfile_16s %>% filter(!startsWith(Study_ID, 'PMC'))
```

The 16s mapfile should have 102 rows and 5 columns. There are 40 FMT samples,
including 28 NoRes and 1 Res. There are 62 Placebo samples, including 58 NoRes
and 4 Res. This is 102 patient samples. No DonorB samples are included in this
mapfile.


```{r}
mapfile_mgm <- read.csv("../data/UCFMT1_METAmap.txt", sep = "\t")
dim(mapfile_mgm)
table(select(mapfile_mgm, Treatment, Remission))
mapfile_mgm %>% filter(!startsWith(Study_ID, 'PMC'))
```

The metagenomics mapfile should have 65 rows and 19 columns. There are 24 FMT
samples, including 12 NoRes and 12 Res. There are 24 Placebo samples, including
20 NoRes samples and 4 Res samples. This makes 48 patient samples. The remaining
17 samples are from DonorB

### Marker files

#### 16S

```{r}
marker_lvl_16s <- read.csv("../data/16s_lvl_RA.csv")
dim(marker_lvl_16s)
```

There are 2945 total 16s markers. This dataframe has 103 columns, which makes
sense. There are 102 samples, plus the marker column.

```{r}
strains16s <- read.csv("../data/jcsz_16s_core_DonB.csv")
dim(strains16s)
head(strains16s)
n_distinct(strains16s)
```

This appears to be a single-column data frame of 16s sequences that appear in
Donor B samples. It has my initials in the file name, but I have no memory of
creating it. I am extremely curious how it was produced. There 457 rows, and
they are unique.

```{r}
donorB_16s_samples <- read.csv("../data/16s_lvl_RA_donorB_samples.csv")
dim(donorB_16s_samples)
colnames(donorB_16s_samples)
any(colnames(donorB_16s_samples)[-1] %in% colnames(marker_lvl_16s))
```

This is a data frame of Donor B sample relative abundances. The DonorB samples
have PMCL sample IDs; however, they are not patient samples because none of them
are in the sample names of the `marker_lvl_16s` data frame.

There are 2945 markers, just like `marker_lvl_16s`, and 35 Donor B samples.

This is used for visualization and benchmarking. It does not come up in the
permutation scripts.

#### Species

```{r}
marker_lvl_sp <- read.csv("../data/species_lvl_RA.csv")
dim(marker_lvl_sp)
colnames(marker_lvl_sp)
```

There are 1087 total species markers. This dataframe has 66 columns, which makes
sense. There are 65 samples, plus the marker column.

##### Strain

```{r}
marker_lvl_st <- read.csv("../data/marker_lvl_RA.csv")
dim(marker_lvl_st)
colnames(marker_lvl_st)
marker_lvl_st %>% select(ends_with('arker')) %>% head()

# check if the Marker column is actually the sp_ and st_ columns merged.
with(marker_lvl_st,
     all(Marker == paste(sp_marker, st_marker, sep = '__')))
```

There are 187,330 total strain markers. This dataframe has 68 columns. In
addition to the 65 sample columns the first column is called `Marker`, and the
last two columns are called `sp_marker` and `st_marker`. The `Marker` column
contains the merged values from `sp_marker` and `st_marker`.

##### MAGs

MAGs were not previously fully processed into a marker_lvl data frame. These are
the input files that were used to do that processing.

```{r}
c_inBin <- read.csv("../data/Assembly/contigs_inBins.txt", sep = "\t", 
                    header = FALSE)
c_inMAG <- read.csv("../data/Assembly/contigs_inMAGs.txt", sep = "\t", 
                    header = FALSE)

dim(c_inBin)
head(c_inBin)
dim(c_inMAG)
head(c_inMAG)

test_MAG = (c_inBin
            %>% filter(V2 %in% c_inMAG$V2))
dim(test_MAG)
head(test_MAG)

all.equal(arrange(test_MAG, V1), arrange(c_inMAG, V1))
```

The `c_inBin` and `c_inMAG` data frames have two columns each and 46,318 and
21,189 rows, respectively. The first column describes a contig ID. The second
column identifies which bin the contig was grouped into. `c_inMAG` is a proper
subset of `c_inBin`.

I have moved the processing to [prep_data.R](./prep_data.R). The marker_lvl data
frame is saved in
[../processed_data/marker_lvl_mg.rds](../processed_data/marker_lvl_mg.rds).

```{r}
load('../processed_data/marker_lvl_mg.rds')
dim(marker_lvl_mg)
colnames(marker_lvl_mg)
```

The `marker_lvl_mg` data frame has 209 markers and 49 columns. This is 48 sample
columns, plus the Marker column. Donor B samples are not included.

##### Genes

```{r}
marker_lvl_ge <- read.csv("../data/genes_lvl.csv")
dim(marker_lvl_ge)
colnames(marker_lvl_ge)
```

The `marker_lvl_ge` data frame has 755,662 rows and 49 columns. This is 48
sample columns, plus the Marker column. DonorB samples are not included.


That's all the input data files. Everything else is processed from them. Code to
do that processing lives in [prep_data.R](./prep_data.R), and output of that
processing is below.

Engraftment Objects
-------------------

### 16S

```{r}
load('../processed_data/eng_16s.rds')
dim(engr_16s)
length(strains16s)
```

According to the engraftment functions, there are 311 engrafted 16s ASVs, which
is not the same as the 313 we saw when we were first working on this. I'm not
thrilled about that.

There are 457 ASVs that are in the donorB samples.

```{r}
profile_16s = get_profile_pat(long_mark_16s, mapfile_16s, cutoff_abs_16S, cutoff_pres_16S,
                       Treatment %in% c('FMT', 'Placebo'))
# Count the number of rows of this data frame contain the word "Engraft"
ct_eng_16s = apply(profile_16s, 1, function(x) 'Engraft' %in% x)
sum(ct_eng_16s)
dim(profile_16s)
```

There are 311 engrafted ASVs, and 457 total ASVs in donor B, which matches

```{r}
profile_16s_don = get_profile_don(long_mark_16s, mapfile_16s, cutoff_abs_16S,
                                  cutoff_pres_16S, Treatment %in% c('FMT', 'Placebo'))
sum(profile_16s_don$status == 'Engraft')
n_distinct(profile_16s_don$Marker)
```

Stil 311 and 457. Great.

### Species

```{r}
load('../processed_data/eng_sp.rds')
dim(engr_sp)
length(strains_sp)
```

According to the engraftment functions, there are 284 engrafted species, and 439
species found in donor B.

```{r}
profile_sp = get_profile_pat(long_mark_sp, mapfile_mgm, cutoff_abs_sp, cutoff_pres_sp,
                       Treatment %in% c('FMT', 'Placebo'))
ct_eng_sp = apply(profile_sp, 1, function(x) 'Engraft' %in% x)
sum(ct_eng_sp)
dim(profile_sp)
```

284 engrafted species and 439 total species in donor B. This agrees with the
engraftment functions

```{r}
profile_sp_don = get_profile_don(long_mark_sp, mapfile_mgm, cutoff_abs_sp,
                                  cutoff_pres_sp, Treatment %in% c('FMT', 'Placebo'))
sum(profile_sp_don$status == 'Engraft')
n_distinct(profile_sp_don$Marker)
```

Still 284 and 439. Good.

### Strains

```{r}
load('../processed_data/eng_st.rds')
dim(engr_st)
length(strains_st)
```

According to the engraftment functions, there are 43,924 engrafted strains, and
a total of 68,162 strains in Donor B samples.

```{r}
profile_st = get_profile_pat(long_mark_st, mapfile_mgm, cutoff_abs_st, cutoff_pres_st,
                       Treatment %in% c('FMT', 'Placebo'))
ct_eng_st = apply(profile_st, 1, function(x) 'Engraft' %in% x)
sum(ct_eng_st)
dim(profile_st)
```

43,924 and 68,162 matches the engraftment functions.

```{r}
profile_st_don = get_profile_don(long_mark_st, mapfile_mgm, cutoff_abs_st,
                                  cutoff_pres_st, Treatment %in% c('FMT', 'Placebo'))
sum(profile_st_don$status == 'Engraft')
n_distinct(profile_st_don$Marker)
```

Still 43,924 and 68,162. Good.

### MAGs

```{r}
load('../processed_data/eng_mg.rds')
dim(engr_mg)
n_distinct(marker_lvl_mg_all$bin_id)
```

According to the engraftment functions, 74 MAGs were engrafted anywhere, and
209 MAGs in Donor B.

```{r}
long_mark_mg = (long_mark_mg
                %>% mutate(abundance = abundance/100))
cutoff_pres_mg = cutoff_pres_mg/100
cutoff_abs_mg = cutoff_abs_mg/100
profile_mg = get_profile_pat(long_mark_mg, mapfile_mgm, cutoff_abs_mg, cutoff_pres_mg,
                       Treatment %in% c('FMT', 'Placebo'))
ct_eng_mg = apply(profile_mg, 1, function(x) 'Engraft' %in% x)
sum(ct_eng_mg)
dim(profile_mg)
```

74 and 209 matches engraftment functions. Good.

```{r}
profile_mg_don = get_profile_don(long_mark_mg, mapfile_mgm, cutoff_abs_mg, cutoff_pres_mg,
                                 Treatment %in% c('FMT', 'Placebo'))
sum(profile_mg_don$status == 'Engraft')
n_distinct(profile_mg_don$Marker)
```

Still 74 and 209. Great.

### Genes

```{r}
load('../processed_data/eng_ge.rds')
dim(engr_ge)
n_distinct((long_mark_ge$Marker))
```

According to engraftment functions, there are 354,728 engrafted genes, and
755,662 total genes in Donor B.

```{r}
long_mark_ge = (long_mark_ge
                %>% mutate(abundance = abundance/100))
cutoff_pres_ge = cutoff_pres_ge/100
cutoff_abs_ge = cutoff_abs_ge/100
profile_ge = get_profile_pat(long_mark_ge, mapfile_mgm, cutoff_abs_ge, cutoff_pres_ge,
                       Treatment %in% c('FMT', 'Placebo'))
ct_eng_ge = apply(profile_ge, 1, function(x) 'Engraft' %in% x)
sum(ct_eng_ge)
dim(profile_ge)
```

This is 354,728 and 755,662, which matches. Good.

```{r}
profile_ge_don = get_profile_don(long_mark_ge, mapfile_mgm, cutoff_abs_ge, cutoff_pres_ge,
                                 Treatment %in% c('FMT','Placebo'))
sum(profile_ge_don$status == 'Engraft')
n_distinct(profile_ge_don$Marker)
```

Still 354,728 and 755,662. Good.